{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28717d89-5c98-462e-a426-c697269ea37f",
   "metadata": {},
   "source": [
    "# International Mouse Phenotyping Consortium (IMPC) Data API Workshop Cheat Sheet\n",
    "Here is a cheat sheet from Data API Workshop.\n",
    "For more information about IMPC visit our website. Useful links:\n",
    "\n",
    "For more information about IMPC visit our [website](https://www.mousephenotype.org/).\n",
    "Other useful links:\n",
    "- Workshop [repository](https://github.com/mpi2/impc-data-api-workshop/tree/main) with all materials\n",
    "- IMPC Solr cores [documentation](https://www.ebi.ac.uk/mi/impc/solrdoc/)\n",
    "- International Mouse Phenotyping Resource of Standardised Screens | [IMPReSS](https://www.mousephenotype.org/impress/index)\n",
    "- The Genome Targeting Repository | [GenTaR](https://www.gentar.org/tracker/#/)\n",
    "\n",
    "Phenodigm useful links:\n",
    "- [IMPC disease models summary](https://www.mousephenotype.org/help/data-visualization/gene-pages/disease-models/)\n",
    "- [IMPC disease associations](https://www.mousephenotype.org/help/data-analysis/disease-associations/)\n",
    "- [Disease Models Portal](https://diseasemodels.research.its.qmul.ac.uk)\n",
    "- [OMIM](https://www.omim.org/)\n",
    "- [Orphanet](https://www.orpha.net)\n",
    "- [SOLR Phenodigm core documentation](https://www.ebi.ac.uk/mi/impc/solrdoc/phenodigm.html)\n",
    "# Set up\n",
    "Let's start! First of all we need to import python libraries and set up helper functions.\n",
    "\n",
    "Execute cell below. Follow steps:\n",
    "1. Select cell by clicking into it.\n",
    "2. Execute code by pressing ▷ play button above.\n",
    "3. You can also use hotkey Ctrl + Enter to execute code.\n",
    "\n",
    "### Helper functions\n",
    "1. `solr_request` — Performs a single Solr request\n",
    "2. `batch_request` — Calls `solr_request` multiple times with `params` to retrieve results in chunk `batch_size` rows at a time.\n",
    "3. `facet_request` — Performs a faceting Solr request\n",
    "4. `entity_iterator` — Generator function fetches results from the Solr server in chunks using pagination.\n",
    "5. `iterator_solr_request` — Fetches results in batches from the Solr API and write them to a file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "20ec2747-a9db-4324-a610-f0a17868c8bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import json\n",
    "from urllib.parse import unquote\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "from IPython.display import display\n",
    "from tqdm import tqdm\n",
    "\n",
    "# Display the whole dataframe <15\n",
    "pd.set_option('display.max_rows', 15)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "# Create helper function\n",
    "def solr_request(core, params, silent=False):\n",
    "    \"\"\"Performs a single Solr request.\n",
    "    \n",
    "    Returns:\n",
    "        num_found: How many rows in total did the request match.\n",
    "        df: A Pandas dataframe with a portion of the request matching `start` and `rows` parameters.\n",
    "        silent: Suppress displaying the df and number of results (useful for batch requests).\n",
    "    \"\"\"\n",
    "    base_url = \"https://www.ebi.ac.uk/mi/impc/solr/\"\n",
    "    solr_url = base_url + core + \"/select\"\n",
    "\n",
    "    response = requests.get(solr_url, params=params)\n",
    "    if not silent:\n",
    "        print(f\"\\nYour request:\\n{response.request.url}\\n\")\n",
    "    \n",
    "    # Check if the request was successful (status code 200)\n",
    "    if response.status_code == 200:\n",
    "        # Parse the JSON response\n",
    "        data = response.json()\n",
    "        num_found = data[\"response\"][\"numFound\"]\n",
    "        if not silent:\n",
    "            print(f'Number of found documents: {num_found}\\n')\n",
    "        # Extract and add search results to the list\n",
    "        search_results = []\n",
    "        for doc in data[\"response\"][\"docs\"]:\n",
    "            search_results.append(doc)\n",
    "    \n",
    "        # Convert the list of dictionaries into a DataFrame and print the DataFrame\n",
    "        df = pd.DataFrame(search_results)\n",
    "        if not silent:\n",
    "            display(df)\n",
    "        return num_found, df\n",
    "    \n",
    "    else:\n",
    "        print(\"Error:\", response.status_code, response.text)\n",
    "\n",
    "def batch_request(core, params, batch_size):\n",
    "    \"\"\"Calls `solr_request` multiple times with `params` to retrieve results in chunk `batch_size` rows at a time.\"\"\"\n",
    "    if \"rows\" in \"params\":\n",
    "        print(\"WARN: You have specified the `params` -> `rows` value. It will be ignored, because the data is retrieved `batch_size` rows at a time.\")\n",
    "    # Determine the total number of rows. Note that we do not request any data (rows = 0).\n",
    "    num_results, _ = solr_request(core=core, params={**params, \"start\": 0, \"rows\": 0}, silent=True)\n",
    "    # Initialise everything for data retrieval.\n",
    "    start = 0\n",
    "    chunks = []\n",
    "    # Request chunks until we have complete data.\n",
    "    with tqdm(total=num_results) as pbar:  # Initialize tqdm progress bar.\n",
    "        while start < num_results:\n",
    "            # Update progress bar with the number of rows requested.\n",
    "            pbar.update(batch_size) \n",
    "            # Request chunk. We don't need num_results anymore because it does not change.\n",
    "            _, df_chunk = solr_request(core=core, params={**params, \"start\": start, \"rows\": batch_size}, silent=True)\n",
    "            # Record chunk.\n",
    "            chunks.append(df_chunk)\n",
    "            # Increment start.\n",
    "            start += batch_size\n",
    "    # Prepare final dataframe.\n",
    "    return pd.concat(chunks, ignore_index=True)\n",
    "\n",
    "def facet_request(core, params, silent=False):\n",
    "    \"\"\"Performs a faceting Solr request.\n",
    "    \n",
    "    Returns:\n",
    "        num_found: How many rows in total did the request match.\n",
    "        df: A Pandas dataframe with a portion of the request matching `start` and `rows` parameters.\n",
    "        silent: Suppress displaying the df and number of results (useful for batch requests).\n",
    "    \"\"\"\n",
    "    base_url = \"https://www.ebi.ac.uk/mi/impc/solr/\"\n",
    "    solr_url = base_url + core + \"/select\"\n",
    "\n",
    "    response = requests.get(solr_url, params=params)\n",
    "    if not silent:\n",
    "        print(f\"\\nYour request:\\n{unquote(response.request.url)}\\n\")\n",
    "    \n",
    "    # Check if the request was successful (status code 200)\n",
    "    if response.status_code == 200:\n",
    "        # Parse the JSON response\n",
    "        data = response.json()\n",
    "        num_found = data[\"response\"][\"numFound\"]\n",
    "        if not silent:\n",
    "            print(f'Number of found documents: {num_found}\\n')\n",
    "        # Extract and add faceting query results to the list\n",
    "        facet_counts = data[\"facet_counts\"][\"facet_fields\"][params[\"facet.field\"]]\n",
    "        # Initialize an empty dictionary\n",
    "        faceting_dict = {}\n",
    "        # Iterate over the list, taking pairs of elements\n",
    "        for i in range(0, len(facet_counts), 2):\n",
    "            # Assign label as key and count as value\n",
    "            label = facet_counts[i]\n",
    "            count = facet_counts[i + 1]\n",
    "            faceting_dict[label] = [count]\n",
    "        \n",
    "        # Print the resulting dictionary\n",
    "        # Convert the list of dictionaries into a DataFrame and print the DataFrame\n",
    "        df = pd.DataFrame(faceting_dict)\n",
    "        df = pd.DataFrame.from_dict(faceting_dict, orient='index', columns=['counts']).reset_index()\n",
    "\n",
    "        # Rename the columns\n",
    "        df.columns = [params[\"facet.field\"], 'count_per_category']\n",
    "        if not silent:\n",
    "            display(df)\n",
    "        return num_found, df\n",
    "    \n",
    "    else:\n",
    "        print(\"Error:\", response.status_code, response.text)\n",
    "\n",
    "# Helper function to fetch results. This function is used by the 'iterator_solr_request' function.\n",
    "def entity_iterator(base_url, params):\n",
    "    \"\"\"Generator function to fetch results from the SOLR server in chunks using pagination\n",
    "\n",
    "    Args:\n",
    "        base_url (str): The base URL of the Solr server to fetch documents from.\n",
    "        params (dict): A dictionary of parameters to include in the GET request. Must include\n",
    "                       'start' and 'rows' keys, which represent the index of the first document\n",
    "                       to fetch and the number of documents to fetch per request, respectively.\n",
    "\n",
    "    Yields:\n",
    "        dict: The next document in the response from the Solr server.\n",
    "    \"\"\"\n",
    "    # Initialise variable to check the first request\n",
    "    first_request = True\n",
    "\n",
    "    # Call the API in chunks and yield the documents in each chunk\n",
    "    while True:\n",
    "        response = requests.get(base_url, params=params)\n",
    "        data = response.json()\n",
    "        docs = data[\"response\"][\"docs\"]\n",
    "\n",
    "        # Print the first request only\n",
    "        if first_request:\n",
    "            print(f'Your first request: {response.url}')\n",
    "            first_request = False\n",
    "\n",
    "        # Yield the documents in the current chunk\n",
    "        for doc in docs:\n",
    "            yield doc\n",
    "\n",
    "        # Check if there are more results to fetch\n",
    "        start = params[\"start\"] + params[\"rows\"]\n",
    "        num_found = data[\"response\"][\"numFound\"]\n",
    "        if start >= num_found:\n",
    "            break\n",
    "\n",
    "        # Update the start parameter for the next request\n",
    "        params[\"start\"] = start\n",
    "\n",
    "    # Print last request and total number of documents retrieved\n",
    "    print(f'Your last request: {response.url}')\n",
    "    print(f'Number of found documents: {data[\"response\"][\"numFound\"]}\\n')\n",
    "\n",
    "# Function to iterate over field list and write results to a file.\n",
    "def iterator_solr_request(core, params, filename='iteration_solr_request', format='json'):\n",
    "    \"\"\"Function to fetch results in batches from the Solr API and write them to a file\n",
    "        Defaults to fetching 5000 rows at a time.\n",
    "\n",
    "    Args:\n",
    "        core (str): The name of the Solr core to fetch results from.\n",
    "        params (dict): A dictionary of parameters to use in the filter query. Must include\n",
    "                       'field_list' and 'field_type' keys, which represent the list of field items (i.e., list of MGI model identifiers)\n",
    "                        to fetch and the type of the field (i.e., model_id) to filter on, respectively.\n",
    "        filename (str): The name of the file to write the results to. Defaults to 'iteration_solr_request'.\n",
    "        format (str): The format of the output file. Can be 'csv' or 'json'. Defaults to 'json'.\n",
    "    \"\"\"\n",
    "\n",
    "    # Validate format\n",
    "    if format not in ['json','csv']:\n",
    "        raise ValueError(\"Invalid format. Please use 'json' or 'csv'\")\n",
    "    \n",
    "    # Base URL\n",
    "    base_url = \"https://www.ebi.ac.uk/mi/impc/solr/\"\n",
    "    solr_url = base_url + core + \"/select\"\n",
    "\n",
    "    # Extract entities_list and entity_type from params\n",
    "    field_list = params.pop(\"field_list\")\n",
    "    field_type = params.pop(\"field_type\")\n",
    "\n",
    "    # Construct the filter query with grouped model IDs\n",
    "    fq = \"{}:({})\".format(\n",
    "        field_type, \" OR \".join(['\"{}\"'.format(id) for id in field_list])\n",
    "    )\n",
    "\n",
    "    # Show users the field and field values they passed to the function\n",
    "    print(\"Queried field:\",fq)\n",
    "    # Set internal params the users should not change\n",
    "    params[\"fq\"] = fq\n",
    "    params[\"wt\"] = 'json'\n",
    "    params[\"start\"]=0 # Start at the first result\n",
    "    params[\"rows\"]=5000 # Fetch results in chunks of 5000\n",
    "\n",
    "\n",
    "    try:\n",
    "        # Fetch results using a generator function\n",
    "        results_generator = entity_iterator(solr_url, params)\n",
    "    except Exception as e:\n",
    "        raise Exception(\"An error occurred while downloading the data: \" + str(e))\n",
    "\n",
    "    # Append extension to the filename\n",
    "    filename = f\"{filename}.{format}\"\n",
    "\n",
    "    try:\n",
    "        # Open the file in write mode\n",
    "        with open(filename, \"w\", newline=\"\") as f:\n",
    "            if format == 'csv':\n",
    "                writer = None\n",
    "                for item in results_generator:\n",
    "                    # Initialize the CSV writer with the keys of the first item as the field names\n",
    "                    if writer is None:\n",
    "                        writer = csv.DictWriter(f, fieldnames=item.keys())\n",
    "                        writer.writeheader()\n",
    "                    # Write the item to the CSV file\n",
    "                    writer.writerow(item)\n",
    "                    # Write to json without loading to memory.\n",
    "            elif format == 'json':\n",
    "                f.write('[')\n",
    "                for i, item in enumerate(results_generator):\n",
    "                    if i != 0:\n",
    "                        f.write(',')\n",
    "                    json.dump(item, f)\n",
    "                f.write(']')\n",
    "    except Exception as e:\n",
    "        raise Exception(\"An error occurred while writing the file: \" + str(e))\n",
    "\n",
    "    print(f\"File {filename} was created.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f993f59b-6508-4d6b-8725-7efa54661c5d",
   "metadata": {},
   "source": [
    "### 1. How to request limited number of documents?\n",
    "Specify `rows` parameter. In this example we request 3 documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6e2e3ba7-d3ea-4193-876a-1a6106b02c6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=%2A%3A%2A&rows=3\n",
      "\n",
      "Number of found documents: 4070676\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_accession_id</th>\n",
       "      <th>allele_name</th>\n",
       "      <th>allele_symbol</th>\n",
       "      <th>colony_id</th>\n",
       "      <th>data_type</th>\n",
       "      <th>doc_id</th>\n",
       "      <th>female_control_count</th>\n",
       "      <th>female_mutant_count</th>\n",
       "      <th>genetic_background</th>\n",
       "      <th>intermediate_mp_term_id</th>\n",
       "      <th>intermediate_mp_term_name</th>\n",
       "      <th>life_stage_acc</th>\n",
       "      <th>life_stage_name</th>\n",
       "      <th>male_control_count</th>\n",
       "      <th>male_mutant_count</th>\n",
       "      <th>marker_accession_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>metadata</th>\n",
       "      <th>metadata_group</th>\n",
       "      <th>mp_term_id_options</th>\n",
       "      <th>mp_term_name_options</th>\n",
       "      <th>parameter_name</th>\n",
       "      <th>parameter_stable_id</th>\n",
       "      <th>parameter_stable_key</th>\n",
       "      <th>phenotyping_center</th>\n",
       "      <th>pipeline_name</th>\n",
       "      <th>pipeline_stable_id</th>\n",
       "      <th>pipeline_stable_key</th>\n",
       "      <th>procedure_name</th>\n",
       "      <th>procedure_stable_id</th>\n",
       "      <th>procedure_stable_key</th>\n",
       "      <th>production_center</th>\n",
       "      <th>project_name</th>\n",
       "      <th>resource_fullname</th>\n",
       "      <th>resource_name</th>\n",
       "      <th>significant</th>\n",
       "      <th>status</th>\n",
       "      <th>strain_accession_id</th>\n",
       "      <th>strain_name</th>\n",
       "      <th>top_level_mp_term_id</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "      <th>zygosity</th>\n",
       "      <th>_version_</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>p_value</th>\n",
       "      <th>statistical_method</th>\n",
       "      <th>classification_tag</th>\n",
       "      <th>female_control_mean</th>\n",
       "      <th>female_effect_size_low_normal_vs_high</th>\n",
       "      <th>female_effect_size_low_vs_normal_high</th>\n",
       "      <th>female_mutant_mean</th>\n",
       "      <th>female_pvalue_low_normal_vs_high</th>\n",
       "      <th>female_pvalue_low_vs_normal_high</th>\n",
       "      <th>genotype_effect_size_low_normal_vs_high</th>\n",
       "      <th>genotype_effect_size_low_vs_normal_high</th>\n",
       "      <th>genotype_pvalue_low_normal_vs_high</th>\n",
       "      <th>genotype_pvalue_low_vs_normal_high</th>\n",
       "      <th>male_control_mean</th>\n",
       "      <th>male_effect_size_low_normal_vs_high</th>\n",
       "      <th>male_effect_size_low_vs_normal_high</th>\n",
       "      <th>male_mutant_mean</th>\n",
       "      <th>male_pvalue_low_normal_vs_high</th>\n",
       "      <th>male_pvalue_low_vs_normal_high</th>\n",
       "      <th>phenotype_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:5609345</td>\n",
       "      <td>targeted mutation 1b, Helmholtz Zentrum Muench...</td>\n",
       "      <td>Lypla1&lt;tm1b(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>LYPAB</td>\n",
       "      <td>categorical</td>\n",
       "      <td>45497a3b9c8e1012c85154e64a630ea3</td>\n",
       "      <td>932.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>involves: C57BL/6N</td>\n",
       "      <td>[MP:0004508, MP:0005508, MP:0009250]</td>\n",
       "      <td>[abnormal pectoral girdle bone morphology, abn...</td>\n",
       "      <td>IMPCLS:0005</td>\n",
       "      <td>Early adult</td>\n",
       "      <td>917.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>MGI:1344588</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>[Date Scanner equipment last calibrated = 2012...</td>\n",
       "      <td>de98e6e5ea27114612ddffb277851f6b</td>\n",
       "      <td>[MP:0000149]</td>\n",
       "      <td>[abnormal scapula morphology]</td>\n",
       "      <td>Scapulae</td>\n",
       "      <td>IMPC_XRY_006_001</td>\n",
       "      <td>[2082]</td>\n",
       "      <td>BCM</td>\n",
       "      <td>BCM Pipeline</td>\n",
       "      <td>BCM_001</td>\n",
       "      <td>16</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>[IMPC_XRY_001]</td>\n",
       "      <td>[91]</td>\n",
       "      <td>BCM</td>\n",
       "      <td>[BaSH]</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>False</td>\n",
       "      <td>NotProcessed</td>\n",
       "      <td>MGI:2159965</td>\n",
       "      <td>C57BL/6N</td>\n",
       "      <td>[MP:0005390]</td>\n",
       "      <td>[skeleton phenotype]</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>1797159908006690842</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:6399913</td>\n",
       "      <td>endonuclease-mediated mutation 1, The Centre f...</td>\n",
       "      <td>Osbpl8&lt;em1(IMPC)Tcp&gt;</td>\n",
       "      <td>TCPR1519_AEKV</td>\n",
       "      <td>adult-gross-path</td>\n",
       "      <td>9a533ea2c0b3f78ca3ede1d6ed61f2d9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>involves: C57BL/6NCrl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMPCLS:0005</td>\n",
       "      <td>Early adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:2443807</td>\n",
       "      <td>Osbpl8</td>\n",
       "      <td>[Date of sacrifice = 2021-05-11T00:00:00Z|Equi...</td>\n",
       "      <td>d41d8cd98f00b204e9800998ecf8427e</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Epididymis</td>\n",
       "      <td>IMPC_PAT_025_002</td>\n",
       "      <td>[37087]</td>\n",
       "      <td>TCP</td>\n",
       "      <td>TCP Pipeline</td>\n",
       "      <td>TCP_001</td>\n",
       "      <td>9</td>\n",
       "      <td>Gross Pathology and Tissue Collection</td>\n",
       "      <td>[IMPC_PAT_002]</td>\n",
       "      <td>[777]</td>\n",
       "      <td>TCP</td>\n",
       "      <td>[DTCC]</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>False</td>\n",
       "      <td>Successful</td>\n",
       "      <td>MGI:2683688</td>\n",
       "      <td>C57BL/6NCrl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>1797159908006690820</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:5637089</td>\n",
       "      <td>targeted mutation 1b, Wellcome Trust Sanger In...</td>\n",
       "      <td>Raph1&lt;tm1b(EUCOMM)Wtsi&gt;</td>\n",
       "      <td>MUFZ</td>\n",
       "      <td>unidimensional</td>\n",
       "      <td>eb6c509bb88f29d6327bccaa1b3a3bd7</td>\n",
       "      <td>397.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>involves: C57BL/6N</td>\n",
       "      <td>[MP:0001963, MP:0003878, MP:0004738, MP:0006335]</td>\n",
       "      <td>[abnormal hearing electrophysiology, abnormal ...</td>\n",
       "      <td>IMPCLS:0005</td>\n",
       "      <td>Early adult</td>\n",
       "      <td>409.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MGI:1924550</td>\n",
       "      <td>Raph1</td>\n",
       "      <td>[Anesthetic administration route = Intraperito...</td>\n",
       "      <td>59ce9671ec21ce59375623c5c2626a94</td>\n",
       "      <td>[MP:0004738, MP:0011967, MP:0011968]</td>\n",
       "      <td>[abnormal auditory brainstem response, increas...</td>\n",
       "      <td>12kHz-evoked ABR Threshold</td>\n",
       "      <td>IMPC_ABR_006_001</td>\n",
       "      <td>[18930]</td>\n",
       "      <td>WTSI</td>\n",
       "      <td>MGP Select Pipeline</td>\n",
       "      <td>MGP_001</td>\n",
       "      <td>15</td>\n",
       "      <td>Auditory Brain Stem Response</td>\n",
       "      <td>[IMPC_ABR_001]</td>\n",
       "      <td>[547]</td>\n",
       "      <td>WTSI</td>\n",
       "      <td>[BaSH]</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>False</td>\n",
       "      <td>Successful</td>\n",
       "      <td>MGI:2159965</td>\n",
       "      <td>C57BL/6N</td>\n",
       "      <td>[MP:0005377]</td>\n",
       "      <td>[hearing/vestibular/ear phenotype]</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>1797159908924194829</td>\n",
       "      <td>0.154034</td>\n",
       "      <td>0.066499</td>\n",
       "      <td>Reference Range Plus Test framework; quantile ...</td>\n",
       "      <td>Not significant</td>\n",
       "      <td>17.732997</td>\n",
       "      <td>0.066499</td>\n",
       "      <td>0.128463</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.066499</td>\n",
       "      <td>0.128463</td>\n",
       "      <td>0.027709</td>\n",
       "      <td>0.141439</td>\n",
       "      <td>0.594283</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17.432763</td>\n",
       "      <td>0.026895</td>\n",
       "      <td>0.154034</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[male, female]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  allele_accession_id                                        allele_name  \\\n",
       "0         MGI:5609345  targeted mutation 1b, Helmholtz Zentrum Muench...   \n",
       "1         MGI:6399913  endonuclease-mediated mutation 1, The Centre f...   \n",
       "2         MGI:5637089  targeted mutation 1b, Wellcome Trust Sanger In...   \n",
       "\n",
       "              allele_symbol      colony_id         data_type  \\\n",
       "0  Lypla1<tm1b(EUCOMM)Hmgu>          LYPAB       categorical   \n",
       "1      Osbpl8<em1(IMPC)Tcp>  TCPR1519_AEKV  adult-gross-path   \n",
       "2   Raph1<tm1b(EUCOMM)Wtsi>           MUFZ    unidimensional   \n",
       "\n",
       "                             doc_id  female_control_count  \\\n",
       "0  45497a3b9c8e1012c85154e64a630ea3                 932.0   \n",
       "1  9a533ea2c0b3f78ca3ede1d6ed61f2d9                   NaN   \n",
       "2  eb6c509bb88f29d6327bccaa1b3a3bd7                 397.0   \n",
       "\n",
       "   female_mutant_count     genetic_background  \\\n",
       "0                  5.0     involves: C57BL/6N   \n",
       "1                  NaN  involves: C57BL/6NCrl   \n",
       "2                  5.0     involves: C57BL/6N   \n",
       "\n",
       "                            intermediate_mp_term_id  \\\n",
       "0              [MP:0004508, MP:0005508, MP:0009250]   \n",
       "1                                               NaN   \n",
       "2  [MP:0001963, MP:0003878, MP:0004738, MP:0006335]   \n",
       "\n",
       "                           intermediate_mp_term_name life_stage_acc  \\\n",
       "0  [abnormal pectoral girdle bone morphology, abn...    IMPCLS:0005   \n",
       "1                                                NaN    IMPCLS:0005   \n",
       "2  [abnormal hearing electrophysiology, abnormal ...    IMPCLS:0005   \n",
       "\n",
       "  life_stage_name  male_control_count  male_mutant_count marker_accession_id  \\\n",
       "0     Early adult               917.0                5.0         MGI:1344588   \n",
       "1     Early adult                 NaN                NaN         MGI:2443807   \n",
       "2     Early adult               409.0                1.0         MGI:1924550   \n",
       "\n",
       "  marker_symbol                                           metadata  \\\n",
       "0        Lypla1  [Date Scanner equipment last calibrated = 2012...   \n",
       "1        Osbpl8  [Date of sacrifice = 2021-05-11T00:00:00Z|Equi...   \n",
       "2         Raph1  [Anesthetic administration route = Intraperito...   \n",
       "\n",
       "                     metadata_group                    mp_term_id_options  \\\n",
       "0  de98e6e5ea27114612ddffb277851f6b                          [MP:0000149]   \n",
       "1  d41d8cd98f00b204e9800998ecf8427e                                   NaN   \n",
       "2  59ce9671ec21ce59375623c5c2626a94  [MP:0004738, MP:0011967, MP:0011968]   \n",
       "\n",
       "                                mp_term_name_options  \\\n",
       "0                      [abnormal scapula morphology]   \n",
       "1                                                NaN   \n",
       "2  [abnormal auditory brainstem response, increas...   \n",
       "\n",
       "               parameter_name parameter_stable_id parameter_stable_key  \\\n",
       "0                    Scapulae    IMPC_XRY_006_001               [2082]   \n",
       "1                  Epididymis    IMPC_PAT_025_002              [37087]   \n",
       "2  12kHz-evoked ABR Threshold    IMPC_ABR_006_001              [18930]   \n",
       "\n",
       "  phenotyping_center        pipeline_name pipeline_stable_id  \\\n",
       "0                BCM         BCM Pipeline            BCM_001   \n",
       "1                TCP         TCP Pipeline            TCP_001   \n",
       "2               WTSI  MGP Select Pipeline            MGP_001   \n",
       "\n",
       "   pipeline_stable_key                         procedure_name  \\\n",
       "0                   16                                  X-ray   \n",
       "1                    9  Gross Pathology and Tissue Collection   \n",
       "2                   15           Auditory Brain Stem Response   \n",
       "\n",
       "  procedure_stable_id procedure_stable_key production_center project_name  \\\n",
       "0      [IMPC_XRY_001]                 [91]               BCM       [BaSH]   \n",
       "1      [IMPC_PAT_002]                [777]               TCP       [DTCC]   \n",
       "2      [IMPC_ABR_001]                [547]              WTSI       [BaSH]   \n",
       "\n",
       "  resource_fullname resource_name  significant        status  \\\n",
       "0              IMPC          IMPC        False  NotProcessed   \n",
       "1              IMPC          IMPC        False    Successful   \n",
       "2              IMPC          IMPC        False    Successful   \n",
       "\n",
       "  strain_accession_id  strain_name top_level_mp_term_id  \\\n",
       "0         MGI:2159965     C57BL/6N         [MP:0005390]   \n",
       "1         MGI:2683688  C57BL/6NCrl                  NaN   \n",
       "2         MGI:2159965     C57BL/6N         [MP:0005377]   \n",
       "\n",
       "               top_level_mp_term_name    zygosity            _version_  \\\n",
       "0                [skeleton phenotype]  homozygote  1797159908006690842   \n",
       "1                                 NaN  homozygote  1797159908006690820   \n",
       "2  [hearing/vestibular/ear phenotype]  homozygote  1797159908924194829   \n",
       "\n",
       "   effect_size   p_value                                 statistical_method  \\\n",
       "0          NaN       NaN                                                NaN   \n",
       "1     0.000000  1.000000                                   Supplied as data   \n",
       "2     0.154034  0.066499  Reference Range Plus Test framework; quantile ...   \n",
       "\n",
       "  classification_tag  female_control_mean  \\\n",
       "0                NaN                  NaN   \n",
       "1                NaN                  NaN   \n",
       "2    Not significant            17.732997   \n",
       "\n",
       "   female_effect_size_low_normal_vs_high  \\\n",
       "0                                    NaN   \n",
       "1                                    NaN   \n",
       "2                               0.066499   \n",
       "\n",
       "   female_effect_size_low_vs_normal_high  female_mutant_mean  \\\n",
       "0                                    NaN                 NaN   \n",
       "1                                    NaN                 NaN   \n",
       "2                               0.128463                20.0   \n",
       "\n",
       "   female_pvalue_low_normal_vs_high  female_pvalue_low_vs_normal_high  \\\n",
       "0                               NaN                               NaN   \n",
       "1                               NaN                               NaN   \n",
       "2                          0.066499                          0.128463   \n",
       "\n",
       "   genotype_effect_size_low_normal_vs_high  \\\n",
       "0                                      NaN   \n",
       "1                                      NaN   \n",
       "2                                 0.027709   \n",
       "\n",
       "   genotype_effect_size_low_vs_normal_high  \\\n",
       "0                                      NaN   \n",
       "1                                      NaN   \n",
       "2                                 0.141439   \n",
       "\n",
       "   genotype_pvalue_low_normal_vs_high  genotype_pvalue_low_vs_normal_high  \\\n",
       "0                                 NaN                                 NaN   \n",
       "1                                 NaN                                 NaN   \n",
       "2                            0.594283                                 1.0   \n",
       "\n",
       "   male_control_mean  male_effect_size_low_normal_vs_high  \\\n",
       "0                NaN                                  NaN   \n",
       "1                NaN                                  NaN   \n",
       "2          17.432763                             0.026895   \n",
       "\n",
       "   male_effect_size_low_vs_normal_high  male_mutant_mean  \\\n",
       "0                                  NaN               NaN   \n",
       "1                                  NaN               NaN   \n",
       "2                             0.154034              15.0   \n",
       "\n",
       "   male_pvalue_low_normal_vs_high  male_pvalue_low_vs_normal_high  \\\n",
       "0                             NaN                             NaN   \n",
       "1                             NaN                             NaN   \n",
       "2                             1.0                             1.0   \n",
       "\n",
       "    phenotype_sex  \n",
       "0             NaN  \n",
       "1             NaN  \n",
       "2  [male, female]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': '*:*', # Request all records.\n",
    "        'rows': 3   # Request the first three rows\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64e36f81-3803-414d-b565-6b94bb5c7642",
   "metadata": {},
   "source": [
    "### 2. How to request specific fields?\n",
    "Specify `fl` parameter. In this example we request following fields:\n",
    "- `marker_symbol`\n",
    "- `top_level_mp_term_name`\n",
    "- `effect_size`\n",
    "- `p_value`\n",
    "<br>\n",
    "\n",
    "**Warning:** if you misspell field name, no error will be generated and this field will be silently omitted from the final result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "455bc01b-68de-4fbd-8939-71fe3bd1e0a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=%2A%3A%2A&fl=marker_symbol%2Ctop_level_mp_term_name%2Ceffect_size%2Cp_value&rows=3\n",
      "\n",
      "Number of found documents: 4070676\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lypla1</td>\n",
       "      <td>[skeleton phenotype]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Osbpl8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Raph1</td>\n",
       "      <td>[hearing/vestibular/ear phenotype]</td>\n",
       "      <td>0.154034</td>\n",
       "      <td>0.066499</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  marker_symbol              top_level_mp_term_name  effect_size   p_value\n",
       "0        Lypla1                [skeleton phenotype]          NaN       NaN\n",
       "1        Osbpl8                                 NaN     0.000000  1.000000\n",
       "2         Raph1  [hearing/vestibular/ear phenotype]     0.154034  0.066499"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': '*:*', # Request all records.\n",
    "        'fl': 'marker_symbol,top_level_mp_term_name,effect_size,p_value',\n",
    "        'rows': 3   # Request the first three rows\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bda525e-484f-42a0-8a63-d0f81122e4bf",
   "metadata": {},
   "source": [
    "### 3. How to filter by specific field?\n",
    "Specify *field:value* in the `q` parameter. For example, let's filter out *Gprc6a*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "468b9ce0-9646-487d-9083-f77413dcd3b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=marker_symbol%3AGprc6a&fl=marker_symbol%2Ctop_level_mp_term_name%2Ceffect_size%2Cp_value&rows=3\n",
      "\n",
      "Number of found documents: 702\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gprc6a</td>\n",
       "      <td>[behavior/neurological phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gprc6a</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gprc6a</td>\n",
       "      <td>[homeostasis/metabolism phenotype]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  marker_symbol              top_level_mp_term_name\n",
       "0        Gprc6a   [behavior/neurological phenotype]\n",
       "1        Gprc6a                                 NaN\n",
       "2        Gprc6a  [homeostasis/metabolism phenotype]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': 'marker_symbol:Gprc6a', # Request Gprc6a records.\n",
    "        'fl': 'marker_symbol,top_level_mp_term_name,effect_size,p_value',\n",
    "        'rows': 3   # Request the first three rows\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81de90ee-9855-4da7-a65b-3fdfca532d72",
   "metadata": {},
   "source": [
    "### 4. How to search in range of numbers?\n",
    "Use following syntax:\n",
    "- `field:[* TO 100]` — finds all field values less than or equal to 100.\n",
    "- `field:[100 TO *]` — finds all field values greater than or equal to 100.\n",
    "- `field:[* TO *]` — finds any document with a value between the effective values of -Infinity and +Infinity for that field type.\n",
    "- `field:[0 TO 100]` — finds all field values less than or equal to 100 and greater than or equal to 100.\n",
    "<br>\n",
    "\n",
    "An asterisk * may be used for either or both endpoints to specify an open-ended range query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "db4c154e-f9b4-48a0-bdc3-e116bcad3e3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=p_value%3A%5B0+TO+1e-4%5D&fl=marker_symbol%2Ctop_level_mp_term_name%2Ceffect_size%2Cp_value&rows=3\n",
      "\n",
      "Number of found documents: 45387\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>effect_size</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>p_value</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.715391</td>\n",
       "      <td>Tbx22</td>\n",
       "      <td>7.228019e-08</td>\n",
       "      <td>[behavior/neurological phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.973679</td>\n",
       "      <td>Avpr1a</td>\n",
       "      <td>2.229254e-06</td>\n",
       "      <td>[behavior/neurological phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.962857</td>\n",
       "      <td>Trarg1</td>\n",
       "      <td>1.889197e-09</td>\n",
       "      <td>[behavior/neurological phenotype]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   effect_size marker_symbol       p_value             top_level_mp_term_name\n",
       "0     2.715391         Tbx22  7.228019e-08  [behavior/neurological phenotype]\n",
       "1    -0.973679        Avpr1a  2.229254e-06  [behavior/neurological phenotype]\n",
       "2     1.962857        Trarg1  1.889197e-09  [behavior/neurological phenotype]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': 'p_value:[0 TO 1e-4]', # Request p-value from 0 to 1e-4.\n",
    "        'fl': 'marker_symbol,top_level_mp_term_name,effect_size,p_value',\n",
    "        'rows': 3   # Request the first three rows\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4125dc6f-506e-4943-8a0a-cbfe068cffb9",
   "metadata": {},
   "source": [
    "### 5. How to apply multiple filters together?\n",
    "Search parameters can be combined using logical operators.\n",
    "- To match both conditions, you specify `filter1 AND filter2`\n",
    "- To match any one of the conditions, use specify `filter1 OR filter2`\n",
    "For example, to find documents with marker symbol Gprc6a and p-value less than 1e-4:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9febfa14-eba0-4644-a5cb-1247cf8c07ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=marker_symbol%3AGprc6a+AND+p_value%3A%5B0+TO+1e-4%5D&fl=marker_symbol%2Ctop_level_mp_term_name%2Ceffect_size%2Cp_value&rows=3\n",
      "\n",
      "Number of found documents: 3\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>effect_size</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>p_value</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>Gprc6a</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[cardiovascular system phenotype, growth/size/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.307854</td>\n",
       "      <td>Gprc6a</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>[immune system phenotype, endocrine/exocrine g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.253108</td>\n",
       "      <td>Gprc6a</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>[hematopoietic system phenotype]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   effect_size marker_symbol   p_value  \\\n",
       "0     1.000000        Gprc6a  0.000000   \n",
       "1     2.307854        Gprc6a  0.000024   \n",
       "2     1.253108        Gprc6a  0.000088   \n",
       "\n",
       "                              top_level_mp_term_name  \n",
       "0  [cardiovascular system phenotype, growth/size/...  \n",
       "1  [immune system phenotype, endocrine/exocrine g...  \n",
       "2                   [hematopoietic system phenotype]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': 'marker_symbol:Gprc6a AND p_value:[0 TO 1e-4]', # Request p-value from 0 to 1e-4.\n",
    "        'fl': 'marker_symbol,top_level_mp_term_name,effect_size,p_value',\n",
    "        'rows': 3   # Request the first three rows\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f7b7c5a-b18f-4fc9-9f68-e4c938490ea1",
   "metadata": {},
   "source": [
    "### 6. How to apply multiple filters together?\n",
    "Logical filters can be inverted by using the NOT operator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b9bd467a-c043-4047-b8ac-a81408ae4d2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=NOT+marker_symbol%3AGprc6a&fl=marker_symbol%2Ctop_level_mp_term_name%2Ceffect_size%2Cp_value&rows=3\n",
      "\n",
      "Number of found documents: 4069974\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lypla1</td>\n",
       "      <td>[skeleton phenotype]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Osbpl8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Raph1</td>\n",
       "      <td>[hearing/vestibular/ear phenotype]</td>\n",
       "      <td>0.154034</td>\n",
       "      <td>0.066499</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  marker_symbol              top_level_mp_term_name  effect_size   p_value\n",
       "0        Lypla1                [skeleton phenotype]          NaN       NaN\n",
       "1        Osbpl8                                 NaN     0.000000  1.000000\n",
       "2         Raph1  [hearing/vestibular/ear phenotype]     0.154034  0.066499"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': 'NOT marker_symbol:Gprc6a', # Request Gprc6a records.\n",
    "        'fl': 'marker_symbol,top_level_mp_term_name,effect_size,p_value',\n",
    "        'rows': 3   # Request the first three rows\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3c0ea07-77ae-4ed7-bc00-dc4790dfa1fb",
   "metadata": {},
   "source": [
    "### 7. How to deal with Null values?\n",
    "You can filter out null values by applying this range filter that we have seen before: `field:[* TO *]`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ca3a0108-46d0-4ac3-92cb-d985671ac4f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=p_value%3A%5B%2A+TO+%2A%5D&fl=marker_symbol%2Ctop_level_mp_term_name%2Ceffect_size%2Cp_value&rows=3\n",
      "\n",
      "Number of found documents: 2068440\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>effect_size</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>p_value</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>Osbpl8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.154034</td>\n",
       "      <td>Raph1</td>\n",
       "      <td>0.066499</td>\n",
       "      <td>[hearing/vestibular/ear phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000929</td>\n",
       "      <td>Ccdc50</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[behavior/neurological phenotype]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   effect_size marker_symbol   p_value              top_level_mp_term_name\n",
       "0     0.000000        Osbpl8  1.000000                                 NaN\n",
       "1     0.154034         Raph1  0.066499  [hearing/vestibular/ear phenotype]\n",
       "2     0.000929        Ccdc50  1.000000   [behavior/neurological phenotype]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': 'p_value:[* TO *]', # Request p-value from 0 to 1e-4.\n",
    "        'fl': 'marker_symbol,top_level_mp_term_name,effect_size,p_value',\n",
    "        'rows': 3   # Request the first three rows\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96668c44-ea0a-451b-a7af-ccaa37ed841f",
   "metadata": {},
   "source": [
    "### 8. How to download the data?\n",
    "Use batch_request function to download the data. It retrieves results \n",
    "in several chunks. `batch_size` parameter is a size of chunk, that will be used for getting the data. And then run either `to_json` or `to_csv` to save data as json or as CSV.\n",
    "<br>\n",
    "You can add any query into `batch_request` function.\n",
    "<br>\n",
    "In this example we download late adult data for cardiovascular system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7d6a2c46-40b8-475a-a434-d60a22d8318b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10000it [00:09, 1025.75it/s]                                                                                                                     \n"
     ]
    }
   ],
   "source": [
    "df = batch_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': 'top_level_mp_term_name:\"cardiovascular system phenotype\" AND life_stage_name:\"Late adult\"',\n",
    "        'fl': 'allele_accession_id,life_stage_name,marker_symbol,p_value,parameter_name,parameter_stable_id,phenotyping_center,statistical_method,top_level_mp_term_name',\n",
    "    },\n",
    "    batch_size=1000\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e1aefea1-e6d3-480c-91ee-bd409bac7775",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save dataframe to JSON (lines) format for subsequent work. This will contain a single self contained JSON record per line.\n",
    "df.to_json(\"impc_data.json\", orient=\"records\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7c6567a9-d60b-47d7-9771-70364f50364c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can also save as CSV, but note that fine structure such as lists and nested data will be lost.\n",
    "df.to_csv(\"impc_data.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a9137df-39af-4364-8008-341f196465d5",
   "metadata": {},
   "source": [
    "### 9. How to do faceting query?\n",
    "Execute `facet_request` function to estimate counts types of the categories. Required parameters:\n",
    "- 'rows': '0'\n",
    "- 'facet': 'on'\n",
    "- `facet.field` — specifies field for faceting search.\n",
    "- `facet.limit` — specifies the maximum number of facets for a field that should be returned. \n",
    "- `facet.mincount` — specifies the minimum counts required for a facet field to be included in the response."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cf82e559-4dca-48e5-8543-a69df56be8bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=*:*&rows=0&facet=on&facet.field=zygosity&facet.limit=15&facet.mincount=1\n",
      "\n",
      "Number of found documents: 4070676\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>zygosity</th>\n",
       "      <th>count_per_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>homozygote</td>\n",
       "      <td>2520286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>heterozygote</td>\n",
       "      <td>1435388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hemizygote</td>\n",
       "      <td>101488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>13514</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       zygosity  count_per_category\n",
       "0    homozygote             2520286\n",
       "1  heterozygote             1435388\n",
       "2    hemizygote              101488\n",
       "3      wildtype               13514"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = facet_request(\n",
    "    core='statistical-result',\n",
    "    params={\n",
    "        'q': '*:*',\n",
    "        'rows': 0,\n",
    "        'facet': 'on',\n",
    "        'facet.field': 'zygosity',\n",
    "        'facet.limit': 15,\n",
    "        'facet.mincount': 1\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b279449b-c99e-4b82-b87c-5e31ef9355d4",
   "metadata": {},
   "source": [
    "### 10: How to iterate over a list of genes?\n",
    "`iterator_solr_request` will iterate over list of marker_symbols: Zfp580, Firrm, Gpld1 and Prkdc\n",
    "\n",
    "You can download this data in `csv` or `json`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "898090e4-8cc6-48ed-b03b-171f0211c3d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Queried field: marker_symbol:(\"Zfp580\" OR \"Firrm\" OR \"Gpld1\" OR \"Prkdc\")\n",
      "Your first request: https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=%2A%3A%2A&fl=marker_symbol%2Callele_symbol%2Cparameter_stable_id&fq=marker_symbol%3A%28%22Zfp580%22+OR+%22Firrm%22+OR+%22Gpld1%22+OR+%22Prkdc%22%29&wt=json&start=0&rows=5000\n",
      "Your last request: https://www.ebi.ac.uk/mi/impc/solr/statistical-result/select?q=%2A%3A%2A&fl=marker_symbol%2Callele_symbol%2Cparameter_stable_id&fq=marker_symbol%3A%28%22Zfp580%22+OR+%22Firrm%22+OR+%22Gpld1%22+OR+%22Prkdc%22%29&wt=json&start=0&rows=5000\n",
      "Number of found documents: 2052\n",
      "\n",
      "File marker_symbol.json was created.\n"
     ]
    }
   ],
   "source": [
    "# Genes example\n",
    "genes = [\"Zfp580\",\"Firrm\",\"Gpld1\",\"Prkdc\"]\n",
    "\n",
    "# Initial query parameters\n",
    "params = {\n",
    "    'q': \"*:*\",\n",
    "    'fl': 'marker_symbol,allele_symbol,parameter_stable_id',\n",
    "    'field_list': genes,\n",
    "    'field_type': \"marker_symbol\"\n",
    "}\n",
    "\n",
    "iterator_solr_request(\n",
    "    core='statistical-result',\n",
    "    params=params,\n",
    "    filename='marker_symbol',\n",
    "    format ='json')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfdb42b6-93c7-4e3f-9b00-0c4357a0fa39",
   "metadata": {},
   "source": [
    "### 11: How to iterate over a list of models?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8c6122dc-0fdb-46af-8d05-d3affe72703c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Queried field: model_id:(\"MGI:3587188\" OR \"MGI:3587185\" OR \"MGI:3605874\" OR \"MGI:2668213\")\n",
      "Your first request: https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary&fl=model_id%2Cmarker_id%2Cdisease_id&fq=model_id%3A%28%22MGI%3A3587188%22+OR+%22MGI%3A3587185%22+OR+%22MGI%3A3605874%22+OR+%22MGI%3A2668213%22%29&wt=json&start=0&rows=5000\n",
      "Your last request: https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary&fl=model_id%2Cmarker_id%2Cdisease_id&fq=model_id%3A%28%22MGI%3A3587188%22+OR+%22MGI%3A3587185%22+OR+%22MGI%3A3605874%22+OR+%22MGI%3A2668213%22%29&wt=json&start=5000&rows=5000\n",
      "Number of found documents: 7675\n",
      "\n",
      "File model_ids.csv was created.\n"
     ]
    }
   ],
   "source": [
    "# List of model IDs.\n",
    "models = [\"MGI:3587188\",\"MGI:3587185\",\"MGI:3605874\",\"MGI:2668213\"]\n",
    "\n",
    "# Call iterator function\n",
    "iterator_solr_request(\n",
    "    core='phenodigm', \n",
    "        params = {\n",
    "        'q': 'type:disease_model_summary',  \n",
    "        'fl': 'model_id,marker_id,disease_id',\n",
    "        'field_list': models,\n",
    "        'field_type': 'model_id'\n",
    "    },\n",
    "    filename='model_ids',\n",
    "    format='csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1243b035-b4a9-4d3a-9c60-f71862d5729c",
   "metadata": {},
   "source": [
    "### 12. How to calculate phenodigm score?\n",
    "1. Retrieve all diseases related to the mouse gene *[Nxn](https://www.mousephenotype.org/data/genes/MGI:109331)* (MGI:109331)\n",
    "2. Select the following fields:\n",
    "    - `marker_id`\n",
    "    - `model_id`\n",
    "    - `disease_id`\n",
    "    - `disease_term`\n",
    "    - `disease_model_avg_norm`\n",
    "    - `disease_model_max_norm`\n",
    "    - `association_curated`\n",
    "3. [**Optional**] To select curated gene-disease associations add a filter to set `association_curated` to `true`, to keep predictions only set to `false`. \n",
    "4. Calculate their **Phenodigm scores** by adding the following to your last field:\n",
    "    -  `phenodigm_score:div(sum(disease_model_avg_norm,disease_model_max_norm),2)`\n",
    "5. [**Optional**] Sort the results in descending order by the calculated Phenodigm score by passing the following to the `sort` parameter:\n",
    "    - `div(sum(disease_model_avg_norm,disease_model_max_norm),2)`\n",
    "\n",
    "**HINT**: do not include the text `phenodigm_score:` to the sort parameter as this will produce an error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "43c30b42-a6b9-4ef9-a6d1-9790d94f6558",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary+AND+marker_id%3A%22MGI%3A109331%22+AND+association_curated%3Atrue&fl=marker_id%2Cmodel_id%2Cdisease_id%2Cdisease_term%2Cdisease_model_avg_norm%2Cdisease_model_max_norm%2Cassociation_curated%2Cphenodigm_score%3Adiv%28sum%28disease_model_avg_norm%2Cdisease_model_max_norm%29%2C2%29&sort=div%28sum%28disease_model_avg_norm%2Cdisease_model_max_norm%29%2C2%29+desc&rows=3\n",
      "\n",
      "Number of found documents: 48\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_id</th>\n",
       "      <th>disease_term</th>\n",
       "      <th>model_id</th>\n",
       "      <th>marker_id</th>\n",
       "      <th>disease_model_avg_norm</th>\n",
       "      <th>disease_model_max_norm</th>\n",
       "      <th>association_curated</th>\n",
       "      <th>phenodigm_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OMIM:618529</td>\n",
       "      <td>Robinow Syndrome, Autosomal Recessive 2</td>\n",
       "      <td>MGI:5548389#hom#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>35.03</td>\n",
       "      <td>64.44</td>\n",
       "      <td>True</td>\n",
       "      <td>49.735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ORPHA:1507</td>\n",
       "      <td>Autosomal Recessive Robinow Syndrome</td>\n",
       "      <td>MGI:5548389#hom#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>27.69</td>\n",
       "      <td>70.72</td>\n",
       "      <td>True</td>\n",
       "      <td>49.205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORPHA:1507</td>\n",
       "      <td>Autosomal Recessive Robinow Syndrome</td>\n",
       "      <td>MGI:4881804</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>21.44</td>\n",
       "      <td>70.56</td>\n",
       "      <td>True</td>\n",
       "      <td>46.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    disease_id                             disease_term  \\\n",
       "0  OMIM:618529  Robinow Syndrome, Autosomal Recessive 2   \n",
       "1   ORPHA:1507     Autosomal Recessive Robinow Syndrome   \n",
       "2   ORPHA:1507     Autosomal Recessive Robinow Syndrome   \n",
       "\n",
       "                 model_id   marker_id  disease_model_avg_norm  \\\n",
       "0  MGI:5548389#hom#embryo  MGI:109331                   35.03   \n",
       "1  MGI:5548389#hom#embryo  MGI:109331                   27.69   \n",
       "2             MGI:4881804  MGI:109331                   21.44   \n",
       "\n",
       "   disease_model_max_norm  association_curated  phenodigm_score  \n",
       "0                   64.44                 True           49.735  \n",
       "1                   70.72                 True           49.205  \n",
       "2                   70.56                 True           46.000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='phenodigm',\n",
    "    params={\n",
    "        'q': 'type:disease_model_summary AND marker_id:\"MGI:109331\" AND association_curated:true',\n",
    "        'fl': 'marker_id,model_id,disease_id,disease_term,disease_model_avg_norm,disease_model_max_norm,association_curated,phenodigm_score:div(sum(disease_model_avg_norm,disease_model_max_norm),2)',\n",
    "        'sort':'div(sum(disease_model_avg_norm,disease_model_max_norm),2) desc',\n",
    "        'rows': 3\n",
    "    }\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
