{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b7c8e987-4016-4c94-bdf2-3f1e56ec596b",
   "metadata": {},
   "source": [
    "# International Mouse Phenotyping Consortium (IMPC) Data API Workshop\n",
    "Welcome to our workshop! In this session, we'll guide you through using Apache Solr API to access IMPC data. After that, we will focus on the `phenodigm` core. By the end, you'll confidently construct Solr queries to extract IMPC datasets. Get ready for hands-on exercises and real-world examples to reinforce your skills!\n",
    "\n",
    "For more information about IMPC visit our [website](https://www.mousephenotype.org/).\n",
    "Other useful links:\n",
    "- Workshop [repository](https://github.com/mpi2/impc-data-api-workshop/tree/main) with all materials\n",
    "- IMPC Solr cores [documentation](https://www.ebi.ac.uk/mi/impc/solrdoc/)\n",
    "- International Mouse Phenotyping Resource of Standardised Screens | [IMPReSS](https://www.mousephenotype.org/impress/index)\n",
    "- The Genome Targeting Repository | [GenTaR](https://www.gentar.org/tracker/#/)\n",
    "\n",
    "# Set up\n",
    "Let's start! First of all we need to import python libraries and set up helper function.\n",
    "### Helper functions\n",
    "Execute cell below. Follow steps:\n",
    "1. Select cell by clicking into it.\n",
    "2. Execute code by pressing â–· play button above.\n",
    "3. You can also use hotkey Ctrl + Enter to execute code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "145b2ed3-97e9-4b8b-b780-b102dd6f3afe",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "from tqdm import tqdm\n",
    "from urllib.parse import unquote\n",
    "\n",
    "import csv\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "# Display the whole dataframe <15\n",
    "pd.set_option('display.max_rows', 15)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "# Create helper function\n",
    "def solr_request(core, params, silent=False):\n",
    "    \"\"\"Performs a single Solr request.\n",
    "    \n",
    "    Returns:\n",
    "        num_found: How many rows in total did the request match.\n",
    "        df: A Pandas dataframe with a portion of the request matching `start` and `rows` parameters.\n",
    "        silent: Suppress displaying the df and number of results (useful for batch requests).\n",
    "    \"\"\"\n",
    "    base_url = \"https://www.ebi.ac.uk/mi/impc/solr/\"\n",
    "    solr_url = base_url + core + \"/select\"\n",
    "\n",
    "    response = requests.get(solr_url, params=params)\n",
    "    if not silent:\n",
    "        print(f\"\\nYour request:\\n{response.request.url}\\n\")\n",
    "    \n",
    "    # Check if the request was successful (status code 200)\n",
    "    if response.status_code == 200:\n",
    "        # Parse the JSON response\n",
    "        data = response.json()\n",
    "        num_found = data[\"response\"][\"numFound\"]\n",
    "        if not silent:\n",
    "            print(f'Number of found documents: {num_found}\\n')\n",
    "        # Extract and add search results to the list\n",
    "        search_results = []\n",
    "        for doc in data[\"response\"][\"docs\"]:\n",
    "            search_results.append(doc)\n",
    "    \n",
    "        # Convert the list of dictionaries into a DataFrame and print the DataFrame\n",
    "        df = pd.DataFrame(search_results)\n",
    "        if not silent:\n",
    "            display(df)\n",
    "        return num_found, df\n",
    "    \n",
    "    else:\n",
    "        print(\"Error:\", response.status_code, response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab77d79-8e0b-46e4-bf60-027cc69e5525",
   "metadata": {},
   "source": [
    "### Example query\n",
    "We will use `solr_request` function to access IMPC data using Solr API. Let's run cell below and investigate the result.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "caf57a86-93ad-4694-a3ad-179bf1438519",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=%2A%3A%2A&rows=10&fl=marker_symbol%2Callele_symbol%2Cparameter_stable_id\n",
      "\n",
      "Number of found documents: 67660\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_symbol</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>parameter_stable_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zfp580&lt;tm1b(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>Zfp580</td>\n",
       "      <td>IMPC_EYE_063_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Firrm&lt;tm1b(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>Firrm</td>\n",
       "      <td>IMPC_XRY_012_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gpld1&lt;tm1.1(KOMP)Vlcg&gt;</td>\n",
       "      <td>Gpld1</td>\n",
       "      <td>IMPC_PAT_017_002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mbip&lt;em1(IMPC)J&gt;</td>\n",
       "      <td>Mbip</td>\n",
       "      <td>IMPC_HEM_001_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Dnmt3l&lt;tm1b(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>Dnmt3l</td>\n",
       "      <td>IMPC_DXA_004_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cd300lg&lt;tm1a(KOMP)Wtsi&gt;</td>\n",
       "      <td>Cd300lg</td>\n",
       "      <td>IMPC_CAL_008_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Gbgt1&lt;tm1.1(KOMP)Vlcg&gt;</td>\n",
       "      <td>Gbgt1</td>\n",
       "      <td>IMPC_GRS_011_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Kcne1&lt;tm1b(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>Kcne1</td>\n",
       "      <td>IMPC_ABR_012_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Recql4&lt;tm1.1(KOMP)Vlcg&gt;</td>\n",
       "      <td>Recql4</td>\n",
       "      <td>IMPC_VIA_067_001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Miga2&lt;tm1a(KOMP)Wtsi&gt;</td>\n",
       "      <td>Miga2</td>\n",
       "      <td>M-G-P_016_001_002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              allele_symbol marker_symbol parameter_stable_id\n",
       "0  Zfp580<tm1b(EUCOMM)Hmgu>        Zfp580    IMPC_EYE_063_001\n",
       "1   Firrm<tm1b(EUCOMM)Hmgu>         Firrm    IMPC_XRY_012_001\n",
       "2    Gpld1<tm1.1(KOMP)Vlcg>         Gpld1    IMPC_PAT_017_002\n",
       "3          Mbip<em1(IMPC)J>          Mbip    IMPC_HEM_001_001\n",
       "4  Dnmt3l<tm1b(EUCOMM)Hmgu>        Dnmt3l    IMPC_DXA_004_001\n",
       "5   Cd300lg<tm1a(KOMP)Wtsi>       Cd300lg    IMPC_CAL_008_001\n",
       "6    Gbgt1<tm1.1(KOMP)Vlcg>         Gbgt1    IMPC_GRS_011_001\n",
       "7   Kcne1<tm1b(EUCOMM)Hmgu>         Kcne1    IMPC_ABR_012_001\n",
       "8   Recql4<tm1.1(KOMP)Vlcg>        Recql4    IMPC_VIA_067_001\n",
       "9     Miga2<tm1a(KOMP)Wtsi>         Miga2   M-G-P_016_001_002"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='genotype-phenotype',\n",
    "    params={\n",
    "        'q': '*:*',  # Your query, '*' retrieves all documents\n",
    "        'rows': 10,  # Number of rows to retrieve\n",
    "        'fl': 'marker_symbol,allele_symbol,parameter_stable_id',  # Fields to retrieve\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44c9ca60-9c63-4acc-a1ce-6bc3881fba35",
   "metadata": {},
   "source": [
    "Let's take a look at the output of helper function. You can see following:\n",
    "1. Submitted request, that you can open in browser by clicking into the link.\n",
    "2. Number of documents in the requested dataframe.\n",
    "3. Table with the results of your query. It will display less than 15 rows.\n",
    "\n",
    "Let's get started with the exercises!\n",
    "\n",
    "# Exercise block A\n",
    "\n",
    "### Exercise 1: Getting Familiar with the Core\n",
    "We will be working with `genotype-phenotype` core. To get yourself familiar with data, request 3 rows and all fields from this core.\n",
    "<br>\n",
    "<br>\n",
    "If you complete the exercise successfully, your total number of documents will be **67,660**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "85ddcef9-af7c-45f8-9ec9-4e3da168ab5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=%2A%3A%2A&rows=3\n",
      "\n",
      "Number of found documents: 67660\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_accession_id</th>\n",
       "      <th>allele_name</th>\n",
       "      <th>allele_symbol</th>\n",
       "      <th>assertion_type</th>\n",
       "      <th>assertion_type_id</th>\n",
       "      <th>colony_id</th>\n",
       "      <th>doc_id</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>intermediate_mp_term_id</th>\n",
       "      <th>intermediate_mp_term_name</th>\n",
       "      <th>life_stage_acc</th>\n",
       "      <th>life_stage_name</th>\n",
       "      <th>marker_accession_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>mp_term_id</th>\n",
       "      <th>mp_term_name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parameter_name</th>\n",
       "      <th>parameter_stable_id</th>\n",
       "      <th>parameter_stable_key</th>\n",
       "      <th>phenotyping_center</th>\n",
       "      <th>pipeline_name</th>\n",
       "      <th>pipeline_stable_id</th>\n",
       "      <th>pipeline_stable_key</th>\n",
       "      <th>procedure_name</th>\n",
       "      <th>procedure_stable_id</th>\n",
       "      <th>procedure_stable_key</th>\n",
       "      <th>project_name</th>\n",
       "      <th>resource_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>statistical_method</th>\n",
       "      <th>strain_accession_id</th>\n",
       "      <th>strain_name</th>\n",
       "      <th>top_level_mp_term_id</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "      <th>zygosity</th>\n",
       "      <th>_version_</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:5694173</td>\n",
       "      <td>targeted mutation 1b, Helmholtz Zentrum Muench...</td>\n",
       "      <td>Zfp580&lt;tm1b(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>automatic</td>\n",
       "      <td>ECO:0000203</td>\n",
       "      <td>ZFOFB</td>\n",
       "      <td>1743756722287</td>\n",
       "      <td>-1.766374</td>\n",
       "      <td>[MP:0006069, MP:0002864, MP:0003727, MP:000209...</td>\n",
       "      <td>[abnormal retina neuronal layer morphology, ab...</td>\n",
       "      <td>IMPCLS:0005</td>\n",
       "      <td>Early adult</td>\n",
       "      <td>MGI:1916242</td>\n",
       "      <td>Zfp580</td>\n",
       "      <td>MP:0003733</td>\n",
       "      <td>abnormal retina inner nuclear layer morphology</td>\n",
       "      <td>2.643851e-06</td>\n",
       "      <td>Right inner nuclear layer</td>\n",
       "      <td>IMPC_EYE_063_001</td>\n",
       "      <td>[86016]</td>\n",
       "      <td>BCM</td>\n",
       "      <td>BCM Pipeline</td>\n",
       "      <td>BCM_001</td>\n",
       "      <td>16</td>\n",
       "      <td>Eye Morphology</td>\n",
       "      <td>[IMPC_EYE_003]</td>\n",
       "      <td>[1390]</td>\n",
       "      <td>[BaSH]</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>not_considered</td>\n",
       "      <td>Linear Mixed Model framework, LME, not includi...</td>\n",
       "      <td>MGI:2159965</td>\n",
       "      <td>C57BL/6N</td>\n",
       "      <td>[MP:0005391]</td>\n",
       "      <td>[vision/eye phenotype]</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>1797159224090820608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:5548899</td>\n",
       "      <td>targeted mutation 1b, Helmholtz Zentrum Muench...</td>\n",
       "      <td>Firrm&lt;tm1b(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>automatic</td>\n",
       "      <td>ECO:0000203</td>\n",
       "      <td>H-BC055324-A04-TM1B</td>\n",
       "      <td>2276332667035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MP:0005508, MP:0009250]</td>\n",
       "      <td>[abnormal skeleton morphology, abnormal append...</td>\n",
       "      <td>IMPCLS:0005</td>\n",
       "      <td>Early adult</td>\n",
       "      <td>MGI:3590554</td>\n",
       "      <td>Firrm</td>\n",
       "      <td>MP:0004509</td>\n",
       "      <td>abnormal pelvic girdle bone morphology</td>\n",
       "      <td>5.306344e-08</td>\n",
       "      <td>Pelvis</td>\n",
       "      <td>IMPC_XRY_012_001</td>\n",
       "      <td>[19374]</td>\n",
       "      <td>MRC Harwell</td>\n",
       "      <td>Harwell</td>\n",
       "      <td>HRWL_001</td>\n",
       "      <td>8</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>[IMPC_XRY_001]</td>\n",
       "      <td>[557]</td>\n",
       "      <td>[BaSH]</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>male</td>\n",
       "      <td>Fisher Exact Test framework</td>\n",
       "      <td>MGI:2164831</td>\n",
       "      <td>C57BL/6NTac</td>\n",
       "      <td>[MP:0005390]</td>\n",
       "      <td>[skeleton phenotype]</td>\n",
       "      <td>heterozygote</td>\n",
       "      <td>1797159224118083584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:5695879</td>\n",
       "      <td>targeted mutation 1.1, Velocigene</td>\n",
       "      <td>Gpld1&lt;tm1.1(KOMP)Vlcg&gt;</td>\n",
       "      <td>manual</td>\n",
       "      <td>ECO:0000218</td>\n",
       "      <td>BL3802</td>\n",
       "      <td>2336462209120</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[MP:0031094, MP:0002706, MP:0000516, MP:0002135]</td>\n",
       "      <td>[organomegaly, abnormal kidney size, abnormal ...</td>\n",
       "      <td>IMPCLS:0005</td>\n",
       "      <td>Early adult</td>\n",
       "      <td>MGI:106604</td>\n",
       "      <td>Gpld1</td>\n",
       "      <td>MP:0003068</td>\n",
       "      <td>enlarged kidney</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>IMPC_PAT_017_002</td>\n",
       "      <td>[37205]</td>\n",
       "      <td>UC Davis</td>\n",
       "      <td>UCD Pipeline</td>\n",
       "      <td>UCD_001</td>\n",
       "      <td>13</td>\n",
       "      <td>Gross Pathology and Tissue Collection</td>\n",
       "      <td>[IMPC_PAT_002]</td>\n",
       "      <td>[779]</td>\n",
       "      <td>[DTCC]</td>\n",
       "      <td>IMPC</td>\n",
       "      <td>male</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>MGI:2683688</td>\n",
       "      <td>C57BL/6NCrl</td>\n",
       "      <td>[MP:0005367, MP:0005378]</td>\n",
       "      <td>[renal/urinary system phenotype, growth/size/b...</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>1797159224119132160</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  allele_accession_id                                        allele_name  \\\n",
       "0         MGI:5694173  targeted mutation 1b, Helmholtz Zentrum Muench...   \n",
       "1         MGI:5548899  targeted mutation 1b, Helmholtz Zentrum Muench...   \n",
       "2         MGI:5695879                  targeted mutation 1.1, Velocigene   \n",
       "\n",
       "              allele_symbol assertion_type assertion_type_id  \\\n",
       "0  Zfp580<tm1b(EUCOMM)Hmgu>      automatic       ECO:0000203   \n",
       "1   Firrm<tm1b(EUCOMM)Hmgu>      automatic       ECO:0000203   \n",
       "2    Gpld1<tm1.1(KOMP)Vlcg>         manual       ECO:0000218   \n",
       "\n",
       "             colony_id         doc_id  effect_size  \\\n",
       "0                ZFOFB  1743756722287    -1.766374   \n",
       "1  H-BC055324-A04-TM1B  2276332667035          NaN   \n",
       "2               BL3802  2336462209120     1.000000   \n",
       "\n",
       "                             intermediate_mp_term_id  \\\n",
       "0  [MP:0006069, MP:0002864, MP:0003727, MP:000209...   \n",
       "1                           [MP:0005508, MP:0009250]   \n",
       "2   [MP:0031094, MP:0002706, MP:0000516, MP:0002135]   \n",
       "\n",
       "                           intermediate_mp_term_name life_stage_acc  \\\n",
       "0  [abnormal retina neuronal layer morphology, ab...    IMPCLS:0005   \n",
       "1  [abnormal skeleton morphology, abnormal append...    IMPCLS:0005   \n",
       "2  [organomegaly, abnormal kidney size, abnormal ...    IMPCLS:0005   \n",
       "\n",
       "  life_stage_name marker_accession_id marker_symbol  mp_term_id  \\\n",
       "0     Early adult         MGI:1916242        Zfp580  MP:0003733   \n",
       "1     Early adult         MGI:3590554         Firrm  MP:0004509   \n",
       "2     Early adult          MGI:106604         Gpld1  MP:0003068   \n",
       "\n",
       "                                     mp_term_name       p_value  \\\n",
       "0  abnormal retina inner nuclear layer morphology  2.643851e-06   \n",
       "1          abnormal pelvic girdle bone morphology  5.306344e-08   \n",
       "2                                 enlarged kidney  0.000000e+00   \n",
       "\n",
       "              parameter_name parameter_stable_id parameter_stable_key  \\\n",
       "0  Right inner nuclear layer    IMPC_EYE_063_001              [86016]   \n",
       "1                     Pelvis    IMPC_XRY_012_001              [19374]   \n",
       "2                     Kidney    IMPC_PAT_017_002              [37205]   \n",
       "\n",
       "  phenotyping_center pipeline_name pipeline_stable_id pipeline_stable_key  \\\n",
       "0                BCM  BCM Pipeline            BCM_001                  16   \n",
       "1        MRC Harwell       Harwell           HRWL_001                   8   \n",
       "2           UC Davis  UCD Pipeline            UCD_001                  13   \n",
       "\n",
       "                          procedure_name procedure_stable_id  \\\n",
       "0                         Eye Morphology      [IMPC_EYE_003]   \n",
       "1                                  X-ray      [IMPC_XRY_001]   \n",
       "2  Gross Pathology and Tissue Collection      [IMPC_PAT_002]   \n",
       "\n",
       "  procedure_stable_key project_name resource_name             sex  \\\n",
       "0               [1390]       [BaSH]          IMPC  not_considered   \n",
       "1                [557]       [BaSH]          IMPC            male   \n",
       "2                [779]       [DTCC]          IMPC            male   \n",
       "\n",
       "                                  statistical_method strain_accession_id  \\\n",
       "0  Linear Mixed Model framework, LME, not includi...         MGI:2159965   \n",
       "1                        Fisher Exact Test framework         MGI:2164831   \n",
       "2                                   Supplied as data         MGI:2683688   \n",
       "\n",
       "   strain_name      top_level_mp_term_id  \\\n",
       "0     C57BL/6N              [MP:0005391]   \n",
       "1  C57BL/6NTac              [MP:0005390]   \n",
       "2  C57BL/6NCrl  [MP:0005367, MP:0005378]   \n",
       "\n",
       "                              top_level_mp_term_name      zygosity  \\\n",
       "0                             [vision/eye phenotype]    homozygote   \n",
       "1                               [skeleton phenotype]  heterozygote   \n",
       "2  [renal/urinary system phenotype, growth/size/b...    homozygote   \n",
       "\n",
       "             _version_  \n",
       "0  1797159224090820608  \n",
       "1  1797159224118083584  \n",
       "2  1797159224119132160  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='genotype-phenotype',\n",
    "    params={\n",
    "        'q': '*:*',\n",
    "        'rows': 3,\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75179169-d80e-4d95-bdea-d3a88bb0d576",
   "metadata": {},
   "source": [
    "### Exercise 2: Selecting Specific Fields\n",
    "As you can see, there is a lot of fields. To focus on the fields we need, request only the following once:\n",
    "- marker_symbol\n",
    "- marker_accession_id\n",
    "- zygosity\n",
    "- parameter_name\n",
    "- parameter_stable_id\n",
    "- p_value\n",
    "<br>\n",
    "\n",
    "Modify query from exercise 1 to request limited list of the fields above.\n",
    "<br>\n",
    "<br>\n",
    "If you complete the exercise successfully, your total number of documents will be **67,660**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3b8b77fc-876a-43f7-a09d-ca4e0d2bb439",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=%2A%3A%2A&rows=3&fl=marker_symbol%2Cmarker_accession_id%2Czygosity%2Cparameter_name%2Cparameter_stable_id%2Cp_value\n",
      "\n",
      "Number of found documents: 67660\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>marker_accession_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parameter_name</th>\n",
       "      <th>parameter_stable_id</th>\n",
       "      <th>zygosity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:1916242</td>\n",
       "      <td>Zfp580</td>\n",
       "      <td>2.643851e-06</td>\n",
       "      <td>Right inner nuclear layer</td>\n",
       "      <td>IMPC_EYE_063_001</td>\n",
       "      <td>homozygote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:3590554</td>\n",
       "      <td>Firrm</td>\n",
       "      <td>5.306344e-08</td>\n",
       "      <td>Pelvis</td>\n",
       "      <td>IMPC_XRY_012_001</td>\n",
       "      <td>heterozygote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:106604</td>\n",
       "      <td>Gpld1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>IMPC_PAT_017_002</td>\n",
       "      <td>homozygote</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  marker_accession_id marker_symbol       p_value             parameter_name  \\\n",
       "0         MGI:1916242        Zfp580  2.643851e-06  Right inner nuclear layer   \n",
       "1         MGI:3590554         Firrm  5.306344e-08                     Pelvis   \n",
       "2          MGI:106604         Gpld1  0.000000e+00                     Kidney   \n",
       "\n",
       "  parameter_stable_id      zygosity  \n",
       "0    IMPC_EYE_063_001    homozygote  \n",
       "1    IMPC_XRY_012_001  heterozygote  \n",
       "2    IMPC_PAT_017_002    homozygote  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='genotype-phenotype',\n",
    "    params={\n",
    "        'q': '*:*',\n",
    "        'rows': 3,\n",
    "        'fl': 'marker_symbol,marker_accession_id,zygosity,parameter_name,parameter_stable_id,p_value'\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4e092de-bf4d-42d0-8db2-762f61aea06c",
   "metadata": {},
   "source": [
    "# Exercise block B\n",
    "### Exercise 3: Filtering by Single Field\n",
    "Let's now focus on a particular gene. In this example we will be using *Dclk1*. Filter the results so there only documents of this gene are displayed by modifying query from exercise 2.\n",
    "<br>\n",
    "<br>\n",
    "If you complete the exercise successfully, your total number of documents will be **13**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5843d4fb-5057-468a-8538-85ecbd46c093",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=marker_symbol%3ADclk1&rows=3&fl=marker_symbol%2Cmarker_accession_id%2Czygosity%2Cparameter_name%2Cparameter_stable_id%2Cp_value\n",
      "\n",
      "Number of found documents: 13\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>marker_accession_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parameter_name</th>\n",
       "      <th>parameter_stable_id</th>\n",
       "      <th>zygosity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:1330861</td>\n",
       "      <td>Dclk1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ANA classification</td>\n",
       "      <td>MGP_ANA_002_001</td>\n",
       "      <td>homozygote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:1330861</td>\n",
       "      <td>Dclk1</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>Forelimb grip strength normalised against body...</td>\n",
       "      <td>IMPC_GRS_010_001</td>\n",
       "      <td>heterozygote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1330861</td>\n",
       "      <td>Dclk1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ANA classification</td>\n",
       "      <td>MGP_ANA_002_001</td>\n",
       "      <td>heterozygote</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  marker_accession_id marker_symbol   p_value  \\\n",
       "0         MGI:1330861         Dclk1  1.000000   \n",
       "1         MGI:1330861         Dclk1  0.000006   \n",
       "2         MGI:1330861         Dclk1  1.000000   \n",
       "\n",
       "                                      parameter_name parameter_stable_id  \\\n",
       "0                                 ANA classification     MGP_ANA_002_001   \n",
       "1  Forelimb grip strength normalised against body...    IMPC_GRS_010_001   \n",
       "2                                 ANA classification     MGP_ANA_002_001   \n",
       "\n",
       "       zygosity  \n",
       "0    homozygote  \n",
       "1  heterozygote  \n",
       "2  heterozygote  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='genotype-phenotype',\n",
    "    params={\n",
    "        'q': 'marker_symbol:Dclk1',\n",
    "        'rows': 3,\n",
    "        'fl': 'marker_symbol,marker_accession_id,zygosity,parameter_name,parameter_stable_id,p_value'\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1501dfa9-84e7-4733-b1f7-d92b6e49c935",
   "metadata": {},
   "source": [
    "### Exercise 4: Filtering Numerical Values and Applying Multiple Filters\n",
    "In addition to the `marker_symbol` filter, let's also apply more strict p-value threshold, so that it is less than 1e-4.\n",
    "<br>\n",
    "Modify query from exercise 3 and display **10 rows** instead of 3. \n",
    "<br>\n",
    "Note: Sometimes spelling may differ.\n",
    "<br>e.g. **`p_value`** is the name of the field in Solr, whereas **\"p-value\"** is the term used in real life.\n",
    "<br>\n",
    "<br>\n",
    "If you complete the exercise successfully, your total number of documents will be **5**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f20e3749-73c3-4c1a-a98c-fce067471d88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=marker_symbol%3ADclk1+AND+p_value%3A%5B%2A+TO+1e-4%5D&rows=10&fl=marker_symbol%2Cmarker_accession_id%2Czygosity%2Cparameter_name%2Cparameter_stable_id%2Cp_value\n",
      "\n",
      "Number of found documents: 5\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>marker_accession_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parameter_name</th>\n",
       "      <th>parameter_stable_id</th>\n",
       "      <th>zygosity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:1330861</td>\n",
       "      <td>Dclk1</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>Forelimb grip strength normalised against body...</td>\n",
       "      <td>IMPC_GRS_010_001</td>\n",
       "      <td>heterozygote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:1330861</td>\n",
       "      <td>Dclk1</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>Body length</td>\n",
       "      <td>IMPC_DXA_006_001</td>\n",
       "      <td>homozygote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1330861</td>\n",
       "      <td>Dclk1</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>Magnesium</td>\n",
       "      <td>IMPC_CBC_054_001</td>\n",
       "      <td>homozygote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:1330861</td>\n",
       "      <td>Dclk1</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>Aspartate aminotransferase</td>\n",
       "      <td>IMPC_CBC_012_001</td>\n",
       "      <td>homozygote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:1330861</td>\n",
       "      <td>Dclk1</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>Forelimb grip strength normalised against body...</td>\n",
       "      <td>IMPC_GRS_010_001</td>\n",
       "      <td>homozygote</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  marker_accession_id marker_symbol   p_value  \\\n",
       "0         MGI:1330861         Dclk1  0.000006   \n",
       "1         MGI:1330861         Dclk1  0.000068   \n",
       "2         MGI:1330861         Dclk1  0.000001   \n",
       "3         MGI:1330861         Dclk1  0.000023   \n",
       "4         MGI:1330861         Dclk1  0.000007   \n",
       "\n",
       "                                      parameter_name parameter_stable_id  \\\n",
       "0  Forelimb grip strength normalised against body...    IMPC_GRS_010_001   \n",
       "1                                        Body length    IMPC_DXA_006_001   \n",
       "2                                          Magnesium    IMPC_CBC_054_001   \n",
       "3                         Aspartate aminotransferase    IMPC_CBC_012_001   \n",
       "4  Forelimb grip strength normalised against body...    IMPC_GRS_010_001   \n",
       "\n",
       "       zygosity  \n",
       "0  heterozygote  \n",
       "1    homozygote  \n",
       "2    homozygote  \n",
       "3    homozygote  \n",
       "4    homozygote  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='genotype-phenotype',\n",
    "    params={\n",
    "        'q': 'marker_symbol:Dclk1 AND p_value:[* TO 1e-4]',\n",
    "        'rows': 10,\n",
    "        'fl': 'marker_symbol,marker_accession_id,zygosity,parameter_name,parameter_stable_id,p_value'\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d03cd8ff-2b64-4e4b-b389-9e69c4de29d7",
   "metadata": {},
   "source": [
    "### Exercise 5: Search for parameter stable ID in the IMPReSS and answer questions\n",
    "Follow steps below and answer the questions: \n",
    "1. Navigate to the [IMPReSS website](https://www.mousephenotype.org/impress/index).\n",
    "2. Search for the parameter_stable_id `IMPC_GRS_010_001`, which is referenced in Exercise 4.\n",
    "3. Examine the procedure associated with the pipeline key `IMPC_001` by clicking on the \"View procedure\" button.\n",
    "<br>\n",
    "- Which week was the specimen tested?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a82674b-9b90-4d49-8df1-c5e581d78ac8",
   "metadata": {},
   "source": [
    "# Exercise block C\n",
    "\n",
    "### Exercise 6: Downloading data in chunks\n",
    "We will use `batch_request` function to download data in chunks. Let's execute cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "42a0e635-86d4-4840-82a4-895a76471a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "def batch_request(core, params, batch_size):\n",
    "    \"\"\"Calls `solr_request` multiple times with `params` to retrieve results in chunk `batch_size` rows at a time.\"\"\"\n",
    "    if \"rows\" in \"params\":\n",
    "        print(\"WARN: You have specified the `params` -> `rows` value. It will be ignored, because the data is retrieved `batch_size` rows at a time.\")\n",
    "    # Determine the total number of rows. Note that we do not request any data (rows = 0).\n",
    "    num_results, _ = solr_request(core=core, params={**params, \"start\": 0, \"rows\": 0}, silent=True)\n",
    "    # Initialise everything for data retrieval.\n",
    "    start = 0\n",
    "    chunks = []\n",
    "    # Request chunks until we have complete data.\n",
    "    with tqdm(total=num_results) as pbar:  # Initialize tqdm progress bar.\n",
    "        while start < num_results:\n",
    "            # Update progress bar with the number of rows requested.\n",
    "            pbar.update(batch_size) \n",
    "            # Request chunk. We don't need num_results anymore because it does not change.\n",
    "            _, df_chunk = solr_request(core=core, params={**params, \"start\": start, \"rows\": batch_size}, silent=True)\n",
    "            # Record chunk.\n",
    "            chunks.append(df_chunk)\n",
    "            # Increment start.\n",
    "            start += batch_size\n",
    "    # Prepare final dataframe.\n",
    "    return pd.concat(chunks, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb2f7cee-a86c-4725-996a-f84424bff2b9",
   "metadata": {},
   "source": [
    "First of all, let's construct a query for cardiovascular system. In the example below we request data with following conditions:\n",
    "- `top_level_mp_term_name` field with `cardiovascular system phenotype`.\n",
    "- `effect_size` is not null.\n",
    "- `life_stage_name` is late adult.\n",
    "<br>\n",
    "<br>\n",
    "\n",
    "Run `solr_request` function below and look at the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cdf433f4-d402-424d-92dc-863bd5931edd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=top_level_mp_term_name%3A%22cardiovascular+system+phenotype%22+AND+effect_size%3A%5B%2A+TO+%2A%5D+AND+life_stage_name%3A%22Late+adult%22&fl=allele_accession_id%2Clife_stage_name%2Cmarker_symbol%2Cmp_term_name%2Cp_value%2Cparameter_name%2Cparameter_stable_id%2Cphenotyping_center%2Cstatistical_method%2Ctop_level_mp_term_name%2Ceffect_size\n",
      "\n",
      "Number of found documents: 413\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_accession_id</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>life_stage_name</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>mp_term_name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>parameter_name</th>\n",
       "      <th>parameter_stable_id</th>\n",
       "      <th>phenotyping_center</th>\n",
       "      <th>statistical_method</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:6276919</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Sstr2</td>\n",
       "      <td>enlarged heart</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Heart</td>\n",
       "      <td>UCDLA_PAT_006_002</td>\n",
       "      <td>UC Davis</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>[cardiovascular system phenotype, growth/size/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:6257614</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Etv6</td>\n",
       "      <td>enlarged heart</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Heart</td>\n",
       "      <td>RBRCLA_PAT_006_002</td>\n",
       "      <td>RBRC</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>[cardiovascular system phenotype, growth/size/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:6341961</td>\n",
       "      <td>0.254275</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Ik</td>\n",
       "      <td>abnormal retina blood vessel morphology</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>Retinal Blood Vessels Structure</td>\n",
       "      <td>KMPCLA_EYE_025_001</td>\n",
       "      <td>KMPC</td>\n",
       "      <td>Fisher Exact Test framework</td>\n",
       "      <td>[vision/eye phenotype, cardiovascular system p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:6152521</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Naa10</td>\n",
       "      <td>abnormal heart morphology</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Heart</td>\n",
       "      <td>UCDLA_PAT_006_002</td>\n",
       "      <td>UC Davis</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>[cardiovascular system phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:5755068</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Ptp4a1</td>\n",
       "      <td>abnormal heart morphology</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Heart</td>\n",
       "      <td>TCPLA_PAT_006_002</td>\n",
       "      <td>TCP</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>[cardiovascular system phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MGI:6277003</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Slc9b1</td>\n",
       "      <td>abnormal heart morphology</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Heart</td>\n",
       "      <td>UCDLA_PAT_006_002</td>\n",
       "      <td>UC Davis</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>[cardiovascular system phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MGI:5614615</td>\n",
       "      <td>-1.177116</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Thra</td>\n",
       "      <td>abnormal heart left ventricle morphology</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>LVIDs</td>\n",
       "      <td>ICSLA_ECH_011_001</td>\n",
       "      <td>ICS</td>\n",
       "      <td>Linear Model Using Generalized Least Squares f...</td>\n",
       "      <td>[cardiovascular system phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MGI:6277006</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Crlf2</td>\n",
       "      <td>small heart</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Heart</td>\n",
       "      <td>UCDLA_PAT_006_002</td>\n",
       "      <td>UC Davis</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>[cardiovascular system phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MGI:6277048</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Lrrtm2</td>\n",
       "      <td>abnormal heart morphology</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Heart</td>\n",
       "      <td>UCDLA_PAT_006_002</td>\n",
       "      <td>UC Davis</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>[cardiovascular system phenotype]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MGI:6336212</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Late adult</td>\n",
       "      <td>Gnptab</td>\n",
       "      <td>enlarged heart</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Heart</td>\n",
       "      <td>UCDLA_PAT_006_002</td>\n",
       "      <td>UC Davis</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>[cardiovascular system phenotype, growth/size/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  allele_accession_id  effect_size life_stage_name marker_symbol  \\\n",
       "0         MGI:6276919     1.000000      Late adult         Sstr2   \n",
       "1         MGI:6257614     1.000000      Late adult          Etv6   \n",
       "2         MGI:6341961     0.254275      Late adult            Ik   \n",
       "3         MGI:6152521     1.000000      Late adult         Naa10   \n",
       "4         MGI:5755068     1.000000      Late adult        Ptp4a1   \n",
       "5         MGI:6277003     1.000000      Late adult        Slc9b1   \n",
       "6         MGI:5614615    -1.177116      Late adult          Thra   \n",
       "7         MGI:6277006     1.000000      Late adult         Crlf2   \n",
       "8         MGI:6277048     1.000000      Late adult        Lrrtm2   \n",
       "9         MGI:6336212     1.000000      Late adult        Gnptab   \n",
       "\n",
       "                               mp_term_name   p_value  \\\n",
       "0                            enlarged heart  0.000000   \n",
       "1                            enlarged heart  0.000000   \n",
       "2   abnormal retina blood vessel morphology  0.000038   \n",
       "3                 abnormal heart morphology  0.000000   \n",
       "4                 abnormal heart morphology  0.000000   \n",
       "5                 abnormal heart morphology  0.000000   \n",
       "6  abnormal heart left ventricle morphology  0.000048   \n",
       "7                               small heart  0.000000   \n",
       "8                 abnormal heart morphology  0.000000   \n",
       "9                            enlarged heart  0.000000   \n",
       "\n",
       "                    parameter_name parameter_stable_id phenotyping_center  \\\n",
       "0                            Heart   UCDLA_PAT_006_002           UC Davis   \n",
       "1                            Heart  RBRCLA_PAT_006_002               RBRC   \n",
       "2  Retinal Blood Vessels Structure  KMPCLA_EYE_025_001               KMPC   \n",
       "3                            Heart   UCDLA_PAT_006_002           UC Davis   \n",
       "4                            Heart   TCPLA_PAT_006_002                TCP   \n",
       "5                            Heart   UCDLA_PAT_006_002           UC Davis   \n",
       "6                            LVIDs   ICSLA_ECH_011_001                ICS   \n",
       "7                            Heart   UCDLA_PAT_006_002           UC Davis   \n",
       "8                            Heart   UCDLA_PAT_006_002           UC Davis   \n",
       "9                            Heart   UCDLA_PAT_006_002           UC Davis   \n",
       "\n",
       "                                  statistical_method  \\\n",
       "0                                   Supplied as data   \n",
       "1                                   Supplied as data   \n",
       "2                        Fisher Exact Test framework   \n",
       "3                                   Supplied as data   \n",
       "4                                   Supplied as data   \n",
       "5                                   Supplied as data   \n",
       "6  Linear Model Using Generalized Least Squares f...   \n",
       "7                                   Supplied as data   \n",
       "8                                   Supplied as data   \n",
       "9                                   Supplied as data   \n",
       "\n",
       "                              top_level_mp_term_name  \n",
       "0  [cardiovascular system phenotype, growth/size/...  \n",
       "1  [cardiovascular system phenotype, growth/size/...  \n",
       "2  [vision/eye phenotype, cardiovascular system p...  \n",
       "3                  [cardiovascular system phenotype]  \n",
       "4                  [cardiovascular system phenotype]  \n",
       "5                  [cardiovascular system phenotype]  \n",
       "6                  [cardiovascular system phenotype]  \n",
       "7                  [cardiovascular system phenotype]  \n",
       "8                  [cardiovascular system phenotype]  \n",
       "9  [cardiovascular system phenotype, growth/size/...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='genotype-phenotype',\n",
    "    params={\n",
    "        'q': 'top_level_mp_term_name:\"cardiovascular system phenotype\" AND effect_size:[* TO *] AND life_stage_name:\"Late adult\"',\n",
    "        'fl': 'allele_accession_id,life_stage_name,marker_symbol,mp_term_name,p_value,parameter_name,parameter_stable_id,phenotyping_center,statistical_method,top_level_mp_term_name,effect_size'\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7a84b44-db06-4d81-9c85-8ba1be079618",
   "metadata": {},
   "source": [
    "Download the data for request above by modifying `batch_request` function below. Set `batch_size` parameter to 100. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "41d76df8-b761-4923-afe4-823d6bee0f41",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "500it [00:00, 695.94it/s]                                                                                                                            \n"
     ]
    }
   ],
   "source": [
    "# Request dataframe in chunks.\n",
    "df = batch_request(\n",
    "    core=\"genotype-phenotype\",\n",
    "    params={\n",
    "        'q': 'top_level_mp_term_name:\"cardiovascular system phenotype\" AND effect_size:[* TO *] AND life_stage_name:\"Late adult\"',\n",
    "        'fl': 'allele_accession_id,life_stage_name,marker_symbol,mp_term_name,p_value,parameter_name,parameter_stable_id,phenotyping_center,statistical_method,top_level_mp_term_name,effect_size'\n",
    "    },\n",
    "    batch_size=100\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "37911fec-918d-451c-a4aa-68dbe6a0745c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save dataframe to JSON (lines) format for subsequent work. This will contain a single self contained JSON record per line.\n",
    "df.to_json(\"impc_data.json\", orient=\"records\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "258fd213-3fd5-476e-968b-39e4d46dbd21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can also save as CSV, but note that fine structure such as lists and nested data will be lost.\n",
    "df.to_csv(\"impc_data.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f69a5608-9b47-431f-b1b8-32a78bbf39f8",
   "metadata": {},
   "source": [
    "# Exercise block D\n",
    "\n",
    "### Exercise 7: Faceting Query\n",
    "We will use `facet_request` function to run faceting query. Let's execute cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9004a23a-c453-497a-8d57-d0de8c5ea6ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def facet_request(core, params, silent=False):\n",
    "    \"\"\"Performs a single Solr request.\n",
    "    \n",
    "    Returns:\n",
    "        num_found: How many rows in total did the request match.\n",
    "        df: A Pandas dataframe with a portion of the request matching `start` and `rows` parameters.\n",
    "        silent: Suppress displaying the df and number of results (useful for batch requests).\n",
    "    \"\"\"\n",
    "    base_url = \"https://www.ebi.ac.uk/mi/impc/solr/\"\n",
    "    solr_url = base_url + core + \"/select\"\n",
    "\n",
    "    response = requests.get(solr_url, params=params)\n",
    "    if not silent:\n",
    "        print(f\"\\nYour request:\\n{unquote(response.request.url)}\\n\")\n",
    "    \n",
    "    # Check if the request was successful (status code 200)\n",
    "    if response.status_code == 200:\n",
    "        # Parse the JSON response\n",
    "        data = response.json()\n",
    "        num_found = data[\"response\"][\"numFound\"]\n",
    "        if not silent:\n",
    "            print(f'Number of found documents: {num_found}\\n')\n",
    "        # Extract and add faceting query results to the list\n",
    "        facet_counts = data[\"facet_counts\"][\"facet_fields\"][params[\"facet.field\"]]\n",
    "        # Initialize an empty dictionary\n",
    "        faceting_dict = {}\n",
    "        # Iterate over the list, taking pairs of elements\n",
    "        for i in range(0, len(facet_counts), 2):\n",
    "            # Assign label as key and count as value\n",
    "            label = facet_counts[i]\n",
    "            count = facet_counts[i + 1]\n",
    "            faceting_dict[label] = [count]\n",
    "        \n",
    "        # Print the resulting dictionary\n",
    "        # Convert the list of dictionaries into a DataFrame and print the DataFrame\n",
    "        df = pd.DataFrame(faceting_dict)\n",
    "        df = pd.DataFrame.from_dict(faceting_dict, orient='index', columns=['counts']).reset_index()\n",
    "\n",
    "        # Rename the columns\n",
    "        df.columns = [params[\"facet.field\"], 'count_per_category']\n",
    "        if not silent:\n",
    "            display(df)\n",
    "        return num_found, df\n",
    "    \n",
    "    else:\n",
    "        print(\"Error:\", response.status_code, response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23025eef-805e-4949-aa21-4aaaad2136e1",
   "metadata": {},
   "source": [
    "In this exercise we will be again querying the whole core. We want to count how many documents there are for each value of the `zygosity` fields. Modify query below to get this information.\n",
    "<br>\n",
    "<br>\n",
    "If you complete the exercise successfully, your total number of documents in homozygote category will be **52,606**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "af132e20-6570-4521-b160-9ac2b1f88aae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=*:*&rows=0&facet=on&facet.field=zygosity&facet.limit=15&facet.mincount=1\n",
      "\n",
      "Number of found documents: 67660\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>zygosity</th>\n",
       "      <th>count_per_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>homozygote</td>\n",
       "      <td>52606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>heterozygote</td>\n",
       "      <td>14312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hemizygote</td>\n",
       "      <td>742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       zygosity  count_per_category\n",
       "0    homozygote               52606\n",
       "1  heterozygote               14312\n",
       "2    hemizygote                 742"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = facet_request(\n",
    "    core='genotype-phenotype',\n",
    "    params={\n",
    "        'q': '*:*',\n",
    "        'rows': 0,\n",
    "        'facet': 'on',\n",
    "        'facet.field': 'zygosity',\n",
    "        'facet.limit': 15,\n",
    "        'facet.mincount': 1\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "508fc9f9-6d1b-4518-89ef-86fb68c0a58c",
   "metadata": {},
   "source": [
    "# Exercises: Phenodigm Core\n",
    "\n",
    "Now we will go over the disease section of the IMPC website. The `phenodigm` core contains all of the information regarding human-mouse disease associations.\n",
    "\n",
    "Some useful links are:\n",
    "- [IMPC disease models summary](https://www.mousephenotype.org/help/data-visualization/gene-pages/disease-models/)\n",
    "- [IMPC disease associations](https://www.mousephenotype.org/help/data-analysis/disease-associations/)\n",
    "- [Disease Models Portal](https://diseasemodels.research.its.qmul.ac.uk)\n",
    "- [OMIM](https://www.omim.org/)\n",
    "- [Orphanet](https://www.orpha.net)\n",
    "- [SOLR Phenodigm core documentation](https://www.ebi.ac.uk/mi/impc/solrdoc/phenodigm.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0e918ce-12dc-4f91-a444-39887474dbff",
   "metadata": {},
   "source": [
    "### Exercise 1: Getting Familiar with the Phenodigm Core\n",
    "Get familiar with the structure of the phenodigm core queries. Write a query to retrieve 5 rows of the `disease_model_summary` type\n",
    "\n",
    "If you complete the exercise successfully, your total number of documents will be **8,444,376**.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f9c30f7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary&rows=5\n",
      "\n",
      "Number of found documents: 8444376\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>disease_id</th>\n",
       "      <th>disease_term</th>\n",
       "      <th>model_id</th>\n",
       "      <th>model_source</th>\n",
       "      <th>model_description</th>\n",
       "      <th>model_genetic_background</th>\n",
       "      <th>marker_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>marker_locus</th>\n",
       "      <th>marker_num_models</th>\n",
       "      <th>disease_model_avg_raw</th>\n",
       "      <th>disease_model_avg_norm</th>\n",
       "      <th>disease_model_max_raw</th>\n",
       "      <th>disease_model_max_norm</th>\n",
       "      <th>association_curated</th>\n",
       "      <th>disease_matched_phenotypes</th>\n",
       "      <th>model_matched_phenotypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>OMIM:227645</td>\n",
       "      <td>Fanconi Anemia, Complementation Group C</td>\n",
       "      <td>MGI:3696039</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Fign&lt;fi&gt;/Fign&lt;fi&gt;</td>\n",
       "      <td>mixed</td>\n",
       "      <td>MGI:1890647</td>\n",
       "      <td>Fign</td>\n",
       "      <td>2:63801852-63928382</td>\n",
       "      <td>5</td>\n",
       "      <td>0.60</td>\n",
       "      <td>26.53</td>\n",
       "      <td>2.45</td>\n",
       "      <td>77.45</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0000568 Microphthalmia, HP:0003214 Prolong...</td>\n",
       "      <td>[MP:0001347 absent lacrimal glands, MP:0013387...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>OMIM:227645</td>\n",
       "      <td>Fanconi Anemia, Complementation Group C</td>\n",
       "      <td>MGI:3696114</td>\n",
       "      <td>MGI</td>\n",
       "      <td>C1galt1&lt;ptl1&gt;/C1galt1&lt;ptl1&gt;</td>\n",
       "      <td>C57BL/6-C1galt1&lt;ptl1&gt;</td>\n",
       "      <td>MGI:2151071</td>\n",
       "      <td>C1galt1</td>\n",
       "      <td>6:7845224-7872042</td>\n",
       "      <td>6</td>\n",
       "      <td>0.64</td>\n",
       "      <td>28.23</td>\n",
       "      <td>2.32</td>\n",
       "      <td>73.31</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0001903 Anemia, HP:0000086 Ectopic kidney,...</td>\n",
       "      <td>[MP:0003179 thrombocytopenia, MP:0000228 abnor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>OMIM:227645</td>\n",
       "      <td>Fanconi Anemia, Complementation Group C</td>\n",
       "      <td>MGI:3696354</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Ercc2&lt;tm3Jhjh&gt;/Ercc2&lt;tm3Jhjh&gt;</td>\n",
       "      <td>involves: 129P2/OlaHsd * C57BL/6 * FVB</td>\n",
       "      <td>MGI:95413</td>\n",
       "      <td>Ercc2</td>\n",
       "      <td>7:19115942-19129619</td>\n",
       "      <td>15</td>\n",
       "      <td>0.53</td>\n",
       "      <td>23.38</td>\n",
       "      <td>2.62</td>\n",
       "      <td>82.88</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0003213 Deficient excision of UV-induced p...</td>\n",
       "      <td>[MP:0010308 decreased tumor latency, MP:000515...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>OMIM:227645</td>\n",
       "      <td>Fanconi Anemia, Complementation Group C</td>\n",
       "      <td>MGI:3696370</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Khdrbs1&lt;tm1Rchd&gt;/Khdrbs1&lt;tm1Rchd&gt;</td>\n",
       "      <td>involves: 129S1/Sv * 129X1/SvJ * C57BL/6</td>\n",
       "      <td>MGI:893579</td>\n",
       "      <td>Khdrbs1</td>\n",
       "      <td>4:129596957-129636096</td>\n",
       "      <td>2</td>\n",
       "      <td>0.37</td>\n",
       "      <td>16.44</td>\n",
       "      <td>2.18</td>\n",
       "      <td>69.07</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0011940 Anterior wedging of T12, HP:000081...</td>\n",
       "      <td>[MP:0000168 abnormal bone marrow development, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>OMIM:227645</td>\n",
       "      <td>Fanconi Anemia, Complementation Group C</td>\n",
       "      <td>MGI:3696775</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Kat6a&lt;tm1Iki&gt;/Kat6a&lt;tm1Iki&gt;</td>\n",
       "      <td>Not Specified</td>\n",
       "      <td>MGI:2442415</td>\n",
       "      <td>Kat6a</td>\n",
       "      <td>8:23349551-23433275</td>\n",
       "      <td>11</td>\n",
       "      <td>0.42</td>\n",
       "      <td>18.49</td>\n",
       "      <td>2.40</td>\n",
       "      <td>75.90</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0001903 Anemia, HP:0001876 Pancytopenia, H...</td>\n",
       "      <td>[MP:0000239 absent common myeloid progenitor c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    type   disease_id  \\\n",
       "0  disease_model_summary  OMIM:227645   \n",
       "1  disease_model_summary  OMIM:227645   \n",
       "2  disease_model_summary  OMIM:227645   \n",
       "3  disease_model_summary  OMIM:227645   \n",
       "4  disease_model_summary  OMIM:227645   \n",
       "\n",
       "                              disease_term     model_id model_source  \\\n",
       "0  Fanconi Anemia, Complementation Group C  MGI:3696039          MGI   \n",
       "1  Fanconi Anemia, Complementation Group C  MGI:3696114          MGI   \n",
       "2  Fanconi Anemia, Complementation Group C  MGI:3696354          MGI   \n",
       "3  Fanconi Anemia, Complementation Group C  MGI:3696370          MGI   \n",
       "4  Fanconi Anemia, Complementation Group C  MGI:3696775          MGI   \n",
       "\n",
       "                   model_description  \\\n",
       "0                  Fign<fi>/Fign<fi>   \n",
       "1        C1galt1<ptl1>/C1galt1<ptl1>   \n",
       "2      Ercc2<tm3Jhjh>/Ercc2<tm3Jhjh>   \n",
       "3  Khdrbs1<tm1Rchd>/Khdrbs1<tm1Rchd>   \n",
       "4        Kat6a<tm1Iki>/Kat6a<tm1Iki>   \n",
       "\n",
       "                   model_genetic_background    marker_id marker_symbol  \\\n",
       "0                                     mixed  MGI:1890647          Fign   \n",
       "1                     C57BL/6-C1galt1<ptl1>  MGI:2151071       C1galt1   \n",
       "2    involves: 129P2/OlaHsd * C57BL/6 * FVB    MGI:95413         Ercc2   \n",
       "3  involves: 129S1/Sv * 129X1/SvJ * C57BL/6   MGI:893579       Khdrbs1   \n",
       "4                             Not Specified  MGI:2442415         Kat6a   \n",
       "\n",
       "            marker_locus  marker_num_models  disease_model_avg_raw  \\\n",
       "0    2:63801852-63928382                  5                   0.60   \n",
       "1      6:7845224-7872042                  6                   0.64   \n",
       "2    7:19115942-19129619                 15                   0.53   \n",
       "3  4:129596957-129636096                  2                   0.37   \n",
       "4    8:23349551-23433275                 11                   0.42   \n",
       "\n",
       "   disease_model_avg_norm  disease_model_max_raw  disease_model_max_norm  \\\n",
       "0                   26.53                   2.45                   77.45   \n",
       "1                   28.23                   2.32                   73.31   \n",
       "2                   23.38                   2.62                   82.88   \n",
       "3                   16.44                   2.18                   69.07   \n",
       "4                   18.49                   2.40                   75.90   \n",
       "\n",
       "   association_curated                         disease_matched_phenotypes  \\\n",
       "0                False  [HP:0000568 Microphthalmia, HP:0003214 Prolong...   \n",
       "1                False  [HP:0001903 Anemia, HP:0000086 Ectopic kidney,...   \n",
       "2                False  [HP:0003213 Deficient excision of UV-induced p...   \n",
       "3                False  [HP:0011940 Anterior wedging of T12, HP:000081...   \n",
       "4                False  [HP:0001903 Anemia, HP:0001876 Pancytopenia, H...   \n",
       "\n",
       "                            model_matched_phenotypes  \n",
       "0  [MP:0001347 absent lacrimal glands, MP:0013387...  \n",
       "1  [MP:0003179 thrombocytopenia, MP:0000228 abnor...  \n",
       "2  [MP:0010308 decreased tumor latency, MP:000515...  \n",
       "3  [MP:0000168 abnormal bone marrow development, ...  \n",
       "4  [MP:0000239 absent common myeloid progenitor c...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='phenodigm',\n",
    "    params={\n",
    "        'q': 'type:disease_model_summary',\n",
    "        'rows': 5\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2c2e890-1556-4a39-859c-f6272e224979",
   "metadata": {},
   "source": [
    "### Exercise 2: Filtering by Disease\n",
    "\n",
    "When filtering by disease, we can use the following fields:\n",
    "- `disease_term`\n",
    "- `disease_id`\n",
    "\n",
    "\n",
    "Refer to the [documentation](https://www.ebi.ac.uk/mi/impc/solrdoc/phenodigm.html) and select the most appropriate `field` to filter for the disease: *[Robinow Syndrome](https://www.omim.org/entry/618529?search=618529&highlight=618529)*\n",
    "\n",
    "How many documents are there on *Robinow Syndrome*?\n",
    "\n",
    "If you complete the exercise successfully, your total number of documents will be **11,655**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d82b5a4d-b20d-48b7-935b-157615355357",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary+AND+disease_term%3A%22Robinow+Syndrome%22&rows=5\n",
      "\n",
      "Number of found documents: 11655\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>disease_id</th>\n",
       "      <th>disease_term</th>\n",
       "      <th>model_id</th>\n",
       "      <th>model_source</th>\n",
       "      <th>model_description</th>\n",
       "      <th>model_genetic_background</th>\n",
       "      <th>marker_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>marker_locus</th>\n",
       "      <th>marker_num_models</th>\n",
       "      <th>disease_model_avg_raw</th>\n",
       "      <th>disease_model_avg_norm</th>\n",
       "      <th>disease_model_max_raw</th>\n",
       "      <th>disease_model_max_norm</th>\n",
       "      <th>association_curated</th>\n",
       "      <th>disease_matched_phenotypes</th>\n",
       "      <th>model_matched_phenotypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2166570</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Ednra&lt;tm1Ywa&gt;/Ednra&lt;tm1Ywa&gt;</td>\n",
       "      <td>129S/SvEv-Ednra&lt;tm1Ywa&gt;</td>\n",
       "      <td>MGI:105923</td>\n",
       "      <td>Ednra</td>\n",
       "      <td>8:78389658-78451081</td>\n",
       "      <td>21</td>\n",
       "      <td>0.78</td>\n",
       "      <td>35.47</td>\n",
       "      <td>2.20</td>\n",
       "      <td>68.15</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0011800 Midface retrusion, HP:0000410 Mixe...</td>\n",
       "      <td>[MP:0010543 aorta tubular hypoplasia, MP:00302...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2166944</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Gli3&lt;Xt-J&gt;/Gli3&lt;Xt-J&gt;</td>\n",
       "      <td>involves: C3H * CD-1</td>\n",
       "      <td>MGI:95729</td>\n",
       "      <td>Gli3</td>\n",
       "      <td>13:15638308-15904611</td>\n",
       "      <td>50</td>\n",
       "      <td>0.84</td>\n",
       "      <td>38.27</td>\n",
       "      <td>2.32</td>\n",
       "      <td>72.06</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0002751 Kyphoscoliosis, HP:0000410 Mixed h...</td>\n",
       "      <td>[MP:0000455 abnormal maxilla morphology, MP:00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2167575</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Ptk2&lt;tm1Imeg&gt;/Ptk2&lt;tm1Imeg&gt;</td>\n",
       "      <td>involves: C57BL/6 * CBA</td>\n",
       "      <td>MGI:95481</td>\n",
       "      <td>Ptk2</td>\n",
       "      <td>15:73076951-73295129</td>\n",
       "      <td>11</td>\n",
       "      <td>0.42</td>\n",
       "      <td>18.91</td>\n",
       "      <td>2.21</td>\n",
       "      <td>68.69</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0010882 Pulmonary valve atresia, HP:001166...</td>\n",
       "      <td>[MP:0011201 abnormal visceral yolk sac cavity ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2167786</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Hoxa2&lt;tm1Ipc&gt;/Hoxa2&lt;tm1Ipc&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MGI:96174</td>\n",
       "      <td>Hoxa2</td>\n",
       "      <td>6:52139397-52141811</td>\n",
       "      <td>19</td>\n",
       "      <td>0.54</td>\n",
       "      <td>24.36</td>\n",
       "      <td>3.10</td>\n",
       "      <td>96.32</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0000202 Orofacial cleft, HP:0000410 Mixed ...</td>\n",
       "      <td>[MP:0008383 enlarged gonial bone, MP:0012786 i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>disease_model_summary</td>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2168259</td>\n",
       "      <td>MGI</td>\n",
       "      <td>Myf5&lt;tm1Jae&gt;/Myf5&lt;tm1Jae&gt;</td>\n",
       "      <td>involves: 129S4/SvJae</td>\n",
       "      <td>MGI:97252</td>\n",
       "      <td>Myf5</td>\n",
       "      <td>10:107318769-107321995</td>\n",
       "      <td>19</td>\n",
       "      <td>0.31</td>\n",
       "      <td>13.88</td>\n",
       "      <td>2.50</td>\n",
       "      <td>77.71</td>\n",
       "      <td>False</td>\n",
       "      <td>[HP:0005011 Mesomelic arm shortening, HP:00009...</td>\n",
       "      <td>[MP:0004672 short ribs, MP:0008148 abnormal st...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    type   disease_id      disease_term     model_id  \\\n",
       "0  disease_model_summary  ORPHA:97360  Robinow Syndrome  MGI:2166570   \n",
       "1  disease_model_summary  ORPHA:97360  Robinow Syndrome  MGI:2166944   \n",
       "2  disease_model_summary  ORPHA:97360  Robinow Syndrome  MGI:2167575   \n",
       "3  disease_model_summary  ORPHA:97360  Robinow Syndrome  MGI:2167786   \n",
       "4  disease_model_summary  ORPHA:97360  Robinow Syndrome  MGI:2168259   \n",
       "\n",
       "  model_source            model_description model_genetic_background  \\\n",
       "0          MGI  Ednra<tm1Ywa>/Ednra<tm1Ywa>  129S/SvEv-Ednra<tm1Ywa>   \n",
       "1          MGI        Gli3<Xt-J>/Gli3<Xt-J>     involves: C3H * CD-1   \n",
       "2          MGI  Ptk2<tm1Imeg>/Ptk2<tm1Imeg>  involves: C57BL/6 * CBA   \n",
       "3          MGI  Hoxa2<tm1Ipc>/Hoxa2<tm1Ipc>    involves: 129S2/SvPas   \n",
       "4          MGI    Myf5<tm1Jae>/Myf5<tm1Jae>    involves: 129S4/SvJae   \n",
       "\n",
       "    marker_id marker_symbol            marker_locus  marker_num_models  \\\n",
       "0  MGI:105923         Ednra     8:78389658-78451081                 21   \n",
       "1   MGI:95729          Gli3    13:15638308-15904611                 50   \n",
       "2   MGI:95481          Ptk2    15:73076951-73295129                 11   \n",
       "3   MGI:96174         Hoxa2     6:52139397-52141811                 19   \n",
       "4   MGI:97252          Myf5  10:107318769-107321995                 19   \n",
       "\n",
       "   disease_model_avg_raw  disease_model_avg_norm  disease_model_max_raw  \\\n",
       "0                   0.78                   35.47                   2.20   \n",
       "1                   0.84                   38.27                   2.32   \n",
       "2                   0.42                   18.91                   2.21   \n",
       "3                   0.54                   24.36                   3.10   \n",
       "4                   0.31                   13.88                   2.50   \n",
       "\n",
       "   disease_model_max_norm  association_curated  \\\n",
       "0                   68.15                False   \n",
       "1                   72.06                False   \n",
       "2                   68.69                False   \n",
       "3                   96.32                False   \n",
       "4                   77.71                False   \n",
       "\n",
       "                          disease_matched_phenotypes  \\\n",
       "0  [HP:0011800 Midface retrusion, HP:0000410 Mixe...   \n",
       "1  [HP:0002751 Kyphoscoliosis, HP:0000410 Mixed h...   \n",
       "2  [HP:0010882 Pulmonary valve atresia, HP:001166...   \n",
       "3  [HP:0000202 Orofacial cleft, HP:0000410 Mixed ...   \n",
       "4  [HP:0005011 Mesomelic arm shortening, HP:00009...   \n",
       "\n",
       "                            model_matched_phenotypes  \n",
       "0  [MP:0010543 aorta tubular hypoplasia, MP:00302...  \n",
       "1  [MP:0000455 abnormal maxilla morphology, MP:00...  \n",
       "2  [MP:0011201 abnormal visceral yolk sac cavity ...  \n",
       "3  [MP:0008383 enlarged gonial bone, MP:0012786 i...  \n",
       "4  [MP:0004672 short ribs, MP:0008148 abnormal st...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='phenodigm',\n",
    "    params={\n",
    "        'q': 'type:disease_model_summary AND disease_term:\"Robinow Syndrome\"',\n",
    "        'rows': 5,\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3395f7e6-2a80-488d-9297-ef45c841d0ba",
   "metadata": {},
   "source": [
    "### Exercise 3: Filtering Mouse Models Related to a Disease\n",
    "Retrieve the first 5 rows related to mouse models associated with *Robinow Syndrome*.\n",
    "\n",
    "Use the following fields to obtain the model information:\n",
    "- `model_id`\n",
    "- `disease_term`\n",
    "- `disease_id`\n",
    "- `marker_id`\n",
    "- `marker_symbol`\n",
    "- `model_source`\n",
    "- `model_genetic_background`\n",
    "\n",
    "If you complete the exercise successfully, your total number of documents will be **11,655** and they will only display **7 fields**.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fc3f0bf1-f211-4166-a497-46cd24574c70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary+AND+disease_term%3A%22Robinow+Syndrome%22&rows=5&fl=model_id%2Cdisease_term%2Cdisease_id%2Cmarker_id%2Cmarker_symbol%2Cmodel_source%2Cmodel_genetic_background\n",
      "\n",
      "Number of found documents: 11655\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_id</th>\n",
       "      <th>disease_term</th>\n",
       "      <th>model_id</th>\n",
       "      <th>model_source</th>\n",
       "      <th>model_genetic_background</th>\n",
       "      <th>marker_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2166570</td>\n",
       "      <td>MGI</td>\n",
       "      <td>129S/SvEv-Ednra&lt;tm1Ywa&gt;</td>\n",
       "      <td>MGI:105923</td>\n",
       "      <td>Ednra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2166944</td>\n",
       "      <td>MGI</td>\n",
       "      <td>involves: C3H * CD-1</td>\n",
       "      <td>MGI:95729</td>\n",
       "      <td>Gli3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2167575</td>\n",
       "      <td>MGI</td>\n",
       "      <td>involves: C57BL/6 * CBA</td>\n",
       "      <td>MGI:95481</td>\n",
       "      <td>Ptk2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2167786</td>\n",
       "      <td>MGI</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MGI:96174</td>\n",
       "      <td>Hoxa2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ORPHA:97360</td>\n",
       "      <td>Robinow Syndrome</td>\n",
       "      <td>MGI:2168259</td>\n",
       "      <td>MGI</td>\n",
       "      <td>involves: 129S4/SvJae</td>\n",
       "      <td>MGI:97252</td>\n",
       "      <td>Myf5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    disease_id      disease_term     model_id model_source  \\\n",
       "0  ORPHA:97360  Robinow Syndrome  MGI:2166570          MGI   \n",
       "1  ORPHA:97360  Robinow Syndrome  MGI:2166944          MGI   \n",
       "2  ORPHA:97360  Robinow Syndrome  MGI:2167575          MGI   \n",
       "3  ORPHA:97360  Robinow Syndrome  MGI:2167786          MGI   \n",
       "4  ORPHA:97360  Robinow Syndrome  MGI:2168259          MGI   \n",
       "\n",
       "  model_genetic_background   marker_id marker_symbol  \n",
       "0  129S/SvEv-Ednra<tm1Ywa>  MGI:105923         Ednra  \n",
       "1     involves: C3H * CD-1   MGI:95729          Gli3  \n",
       "2  involves: C57BL/6 * CBA   MGI:95481          Ptk2  \n",
       "3    involves: 129S2/SvPas   MGI:96174         Hoxa2  \n",
       "4    involves: 129S4/SvJae   MGI:97252          Myf5  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='phenodigm',\n",
    "    params={\n",
    "        'q': 'type:disease_model_summary AND disease_term:\"Robinow Syndrome\"',\n",
    "        'rows': 5,\n",
    "        'fl': 'model_id,disease_term,disease_id,marker_id,marker_symbol,model_source,model_genetic_background',\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74e7e170",
   "metadata": {},
   "source": [
    "## Advanced Exercises"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "470e0283-2e88-4c7a-84b0-cd5358d54802",
   "metadata": {},
   "source": [
    "\n",
    "### Exercise 4: Calculate the Phenodigm score\n",
    "\n",
    "In this exercise, we will calculate the Phenodigm score, sort the results by this score, and filter the documents based on gene-disease curation.\n",
    "\n",
    "1. Retrieve all diseases related to the mouse gene *[Nxn](https://www.mousephenotype.org/data/genes/MGI:109331)* (MGI:109331)\n",
    "2. Select the following fields:\n",
    "    - `marker_id`\n",
    "    - `model_id`\n",
    "    - `disease_id`\n",
    "    - `disease_term`\n",
    "    - `disease_model_avg_norm`\n",
    "    - `disease_model_max_norm`\n",
    "    - `association_curated`\n",
    "3. Filter results to keep only those where `association_curated` is `true`.\n",
    "4. Calculate their **Phenodigm scores** by adding the following to your last field:\n",
    "    -  `phenodigm_score:div(sum(disease_model_avg_norm,disease_model_max_norm),2)`\n",
    "5. Sort the results in descending order by the calculated Phenodigm score by passing the following to the `sort` parameter:\n",
    "    - `div(sum(disease_model_avg_norm,disease_model_max_norm),2)`\n",
    "\n",
    "        **HINT**: do not include the text `phenodigm_score:` to the sort parameter as this will produce an error.\n",
    "\n",
    "If you complete the exercise successfully, your total number of documents will be **48**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a7f83a21-ac7e-42e9-b60c-60bc75c09f1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Your request:\n",
      "https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary+AND+marker_id%3A%22MGI%3A109331%22+AND+association_curated%3Atrue&fl=marker_id%2Cmodel_id%2Cdisease_id%2Cdisease_term%2Cdisease_model_avg_norm%2Cdisease_model_max_norm%2Cassociation_curated%2Cphenodigm_score%3Adiv%28sum%28disease_model_avg_norm%2Cdisease_model_max_norm%29%2C2%29&sort=div%28sum%28disease_model_avg_norm%2Cdisease_model_max_norm%29%2C2%29+desc\n",
      "\n",
      "Number of found documents: 48\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_id</th>\n",
       "      <th>disease_term</th>\n",
       "      <th>model_id</th>\n",
       "      <th>marker_id</th>\n",
       "      <th>disease_model_avg_norm</th>\n",
       "      <th>disease_model_max_norm</th>\n",
       "      <th>association_curated</th>\n",
       "      <th>phenodigm_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OMIM:618529</td>\n",
       "      <td>Robinow Syndrome, Autosomal Recessive 2</td>\n",
       "      <td>MGI:5548389#hom#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>35.03</td>\n",
       "      <td>64.44</td>\n",
       "      <td>True</td>\n",
       "      <td>49.735000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ORPHA:1507</td>\n",
       "      <td>Autosomal Recessive Robinow Syndrome</td>\n",
       "      <td>MGI:5548389#hom#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>27.69</td>\n",
       "      <td>70.72</td>\n",
       "      <td>True</td>\n",
       "      <td>49.205000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORPHA:1507</td>\n",
       "      <td>Autosomal Recessive Robinow Syndrome</td>\n",
       "      <td>MGI:4881804</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>21.44</td>\n",
       "      <td>70.56</td>\n",
       "      <td>True</td>\n",
       "      <td>46.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ORPHA:1507</td>\n",
       "      <td>Autosomal Recessive Robinow Syndrome</td>\n",
       "      <td>MGI:5548389#het#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>14.62</td>\n",
       "      <td>70.72</td>\n",
       "      <td>True</td>\n",
       "      <td>42.670002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OMIM:618529</td>\n",
       "      <td>Robinow Syndrome, Autosomal Recessive 2</td>\n",
       "      <td>MGI:5620177#hom#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>15.99</td>\n",
       "      <td>64.44</td>\n",
       "      <td>True</td>\n",
       "      <td>40.215000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OMIM:618529</td>\n",
       "      <td>Robinow Syndrome, Autosomal Recessive 2</td>\n",
       "      <td>MGI:5548389#het#early</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>24.02</td>\n",
       "      <td>56.29</td>\n",
       "      <td>True</td>\n",
       "      <td>40.155000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ORPHA:1507</td>\n",
       "      <td>Autosomal Recessive Robinow Syndrome</td>\n",
       "      <td>MGI:5548389#het#early</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>21.49</td>\n",
       "      <td>56.65</td>\n",
       "      <td>True</td>\n",
       "      <td>39.070000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OMIM:618529</td>\n",
       "      <td>Robinow Syndrome, Autosomal Recessive 2</td>\n",
       "      <td>MGI:5548389#het#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>22.27</td>\n",
       "      <td>54.93</td>\n",
       "      <td>True</td>\n",
       "      <td>38.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ORPHA:1507</td>\n",
       "      <td>Autosomal Recessive Robinow Syndrome</td>\n",
       "      <td>MGI:5620177#hom#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>13.76</td>\n",
       "      <td>57.27</td>\n",
       "      <td>True</td>\n",
       "      <td>35.515000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OMIM:618529</td>\n",
       "      <td>Robinow Syndrome, Autosomal Recessive 2</td>\n",
       "      <td>MGI:6257797#hom#embryo</td>\n",
       "      <td>MGI:109331</td>\n",
       "      <td>7.76</td>\n",
       "      <td>53.32</td>\n",
       "      <td>True</td>\n",
       "      <td>30.540000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    disease_id                             disease_term  \\\n",
       "0  OMIM:618529  Robinow Syndrome, Autosomal Recessive 2   \n",
       "1   ORPHA:1507     Autosomal Recessive Robinow Syndrome   \n",
       "2   ORPHA:1507     Autosomal Recessive Robinow Syndrome   \n",
       "3   ORPHA:1507     Autosomal Recessive Robinow Syndrome   \n",
       "4  OMIM:618529  Robinow Syndrome, Autosomal Recessive 2   \n",
       "5  OMIM:618529  Robinow Syndrome, Autosomal Recessive 2   \n",
       "6   ORPHA:1507     Autosomal Recessive Robinow Syndrome   \n",
       "7  OMIM:618529  Robinow Syndrome, Autosomal Recessive 2   \n",
       "8   ORPHA:1507     Autosomal Recessive Robinow Syndrome   \n",
       "9  OMIM:618529  Robinow Syndrome, Autosomal Recessive 2   \n",
       "\n",
       "                 model_id   marker_id  disease_model_avg_norm  \\\n",
       "0  MGI:5548389#hom#embryo  MGI:109331                   35.03   \n",
       "1  MGI:5548389#hom#embryo  MGI:109331                   27.69   \n",
       "2             MGI:4881804  MGI:109331                   21.44   \n",
       "3  MGI:5548389#het#embryo  MGI:109331                   14.62   \n",
       "4  MGI:5620177#hom#embryo  MGI:109331                   15.99   \n",
       "5   MGI:5548389#het#early  MGI:109331                   24.02   \n",
       "6   MGI:5548389#het#early  MGI:109331                   21.49   \n",
       "7  MGI:5548389#het#embryo  MGI:109331                   22.27   \n",
       "8  MGI:5620177#hom#embryo  MGI:109331                   13.76   \n",
       "9  MGI:6257797#hom#embryo  MGI:109331                    7.76   \n",
       "\n",
       "   disease_model_max_norm  association_curated  phenodigm_score  \n",
       "0                   64.44                 True        49.735000  \n",
       "1                   70.72                 True        49.205000  \n",
       "2                   70.56                 True        46.000000  \n",
       "3                   70.72                 True        42.670002  \n",
       "4                   64.44                 True        40.215000  \n",
       "5                   56.29                 True        40.155000  \n",
       "6                   56.65                 True        39.070000  \n",
       "7                   54.93                 True        38.600000  \n",
       "8                   57.27                 True        35.515000  \n",
       "9                   53.32                 True        30.540000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_found, df = solr_request(\n",
    "    core='phenodigm',\n",
    "    params={\n",
    "        'q': 'type:disease_model_summary AND marker_id:\"MGI:109331\" AND association_curated:true',\n",
    "        'fl': 'marker_id,model_id,disease_id,disease_term,disease_model_avg_norm,disease_model_max_norm,association_curated,phenodigm_score:div(sum(disease_model_avg_norm,disease_model_max_norm),2)',\n",
    "        'sort':'div(sum(disease_model_avg_norm,disease_model_max_norm),2) desc'\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89acbf3f-2490-4438-b250-90f5efdf9af5",
   "metadata": {},
   "source": [
    "\n",
    "### Exercise 5: Iterate over a list of diseases or models/ genes\n",
    "Now, we will define another helper function, `iterator_solr_request`. This function will request information for a list of values of a given `field`. This is particularly useful when we want to request data for specific models or genes.\n",
    "\n",
    "Execute the helper functions below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3bc86728-aee8-44ac-a73c-4433286e00ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "# Helper function to fetch results. This function is used by the 'iterator_solr_request' function.\n",
    "def entity_iterator(base_url, params):\n",
    "    \"\"\"Generator function to fetch results from the SOLR server in chunks using pagination\n",
    "\n",
    "    Args:\n",
    "        base_url (str): The base URL of the Solr server to fetch documents from.\n",
    "        params (dict): A dictionary of parameters to include in the GET request. Must include\n",
    "                       'start' and 'rows' keys, which represent the index of the first document\n",
    "                       to fetch and the number of documents to fetch per request, respectively.\n",
    "\n",
    "    Yields:\n",
    "        dict: The next document in the response from the Solr server.\n",
    "    \"\"\"\n",
    "    # Initialise variable to check the first request\n",
    "    first_request = True\n",
    "\n",
    "    # Call the API in chunks and yield the documents in each chunk\n",
    "    while True:\n",
    "        response = requests.get(base_url, params=params)\n",
    "        data = response.json()\n",
    "        docs = data[\"response\"][\"docs\"]\n",
    "\n",
    "        # Print the first request only\n",
    "        if first_request:\n",
    "            print(f'Your first request: {response.url}')\n",
    "            first_request = False\n",
    "\n",
    "        # Yield the documents in the current chunk\n",
    "        for doc in docs:\n",
    "            yield doc\n",
    "\n",
    "        # Check if there are more results to fetch\n",
    "        start = params[\"start\"] + params[\"rows\"]\n",
    "        num_found = data[\"response\"][\"numFound\"]\n",
    "        if start >= num_found:\n",
    "            break\n",
    "\n",
    "        # Update the start parameter for the next request\n",
    "        params[\"start\"] = start\n",
    "\n",
    "    # Print last request and total number of documents retrieved\n",
    "    print(f'Your last request: {response.url}')\n",
    "    print(f'Number of found documents: {data[\"response\"][\"numFound\"]}\\n')\n",
    "\n",
    "# Function to iterate over field list and write results to a file.\n",
    "def iterator_solr_request(core, params, filename='iteration_solr_request', format='json'):\n",
    "    \"\"\"Function to fetch results in batches from the Solr API and write them to a file\n",
    "        Defaults to fetching 5000 rows at a time.\n",
    "\n",
    "    Args:\n",
    "        core (str): The name of the Solr core to fetch results from.\n",
    "        params (dict): A dictionary of parameters to use in the filter query. Must include\n",
    "                       'field_list' and 'field_type' keys, which represent the list of field items (i.e., list of MGI model identifiers)\n",
    "                        to fetch and the type of the field (i.e., model_id) to filter on, respectively.\n",
    "        filename (str): The name of the file to write the results to. Defaults to 'iteration_solr_request'.\n",
    "        format (str): The format of the output file. Can be 'csv' or 'json'. Defaults to 'json'.\n",
    "    \"\"\"\n",
    "\n",
    "    # Validate format\n",
    "    if format not in ['json','csv']:\n",
    "        raise ValueError(\"Invalid format. Please use 'json' or 'csv'\")\n",
    "    \n",
    "    # Base URL\n",
    "    base_url = \"https://www.ebi.ac.uk/mi/impc/solr/\"\n",
    "    solr_url = base_url + core + \"/select\"\n",
    "\n",
    "    # Extract entities_list and entity_type from params\n",
    "    field_list = params.pop(\"field_list\")\n",
    "    field_type = params.pop(\"field_type\")\n",
    "\n",
    "    # Construct the filter query with grouped model IDs\n",
    "    fq = \"{}:({})\".format(\n",
    "        field_type, \" OR \".join(['\"{}\"'.format(id) for id in field_list])\n",
    "    )\n",
    "\n",
    "    # Show users the field and field values they passed to the function\n",
    "    print(\"Queried field:\",fq)\n",
    "    # Set internal params the users should not change\n",
    "    params[\"fq\"] = fq\n",
    "    params[\"wt\"] = 'json'\n",
    "    params[\"start\"]=0 # Start at the first result\n",
    "    params[\"rows\"]=5000 # Fetch results in chunks of 5000\n",
    "\n",
    "\n",
    "    try:\n",
    "        # Fetch results using a generator function\n",
    "        results_generator = entity_iterator(solr_url, params)\n",
    "    except Exception as e:\n",
    "        raise Exception(\"An error occurred while downloading the data: \" + str(e))\n",
    "\n",
    "    # Append extension to the filename\n",
    "    filename = f\"{filename}.{format}\"\n",
    "\n",
    "    try:\n",
    "        # Open the file in write mode\n",
    "        with open(filename, \"w\", newline=\"\") as f:\n",
    "            if format == 'csv':\n",
    "                writer = None\n",
    "                for item in results_generator:\n",
    "                    # Initialize the CSV writer with the keys of the first item as the field names\n",
    "                    if writer is None:\n",
    "                        writer = csv.DictWriter(f, fieldnames=item.keys())\n",
    "                        writer.writeheader()\n",
    "                    # Write the item to the CSV file\n",
    "                    writer.writerow(item)\n",
    "                    # Write to json without loading to memory.\n",
    "            elif format == 'json':\n",
    "                f.write('[')\n",
    "                for i, item in enumerate(results_generator):\n",
    "                    if i != 0:\n",
    "                        f.write(',')\n",
    "                    json.dump(item, f)\n",
    "                f.write(']')\n",
    "    except Exception as e:\n",
    "        raise Exception(\"An error occurred while writing the file: \" + str(e))\n",
    "\n",
    "    print(f\"File {filename} was created.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b9a3482-e904-48aa-b6a2-78524639c3d4",
   "metadata": {},
   "source": [
    "Here is how to use it:\n",
    "- New information required from the user:\n",
    "    - `field_list`: A list of field names to be queried in the API. (e.g., MGI ids, OMIM ids, gene ids).\n",
    "    - `field_type`: The type of `field` we are querying. (e.g., `model_id`, `disease_id`, `marker_id`).\n",
    "    - `filename`: The name of the file you wish to save the data in.\n",
    "    - `format`: The format in which you want to save the data (e.g., `json`, `csv`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a01f2005-7a88-4c13-8475-fc987dbe45d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Queried field: model_id:(\"MGI:3587188\" OR \"MGI:3587185\" OR \"MGI:3605874\" OR \"MGI:2668213\")\n",
      "Your first request: https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary&fl=model_id%2Cmarker_id%2Cdisease_id&fq=model_id%3A%28%22MGI%3A3587188%22+OR+%22MGI%3A3587185%22+OR+%22MGI%3A3605874%22+OR+%22MGI%3A2668213%22%29&wt=json&start=0&rows=5000\n",
      "Your last request: https://www.ebi.ac.uk/mi/impc/solr/phenodigm/select?q=type%3Adisease_model_summary&fl=model_id%2Cmarker_id%2Cdisease_id&fq=model_id%3A%28%22MGI%3A3587188%22+OR+%22MGI%3A3587185%22+OR+%22MGI%3A3605874%22+OR+%22MGI%3A2668213%22%29&wt=json&start=5000&rows=5000\n",
      "Number of found documents: 7675\n",
      "\n",
      "File model_ids.csv was created.\n"
     ]
    }
   ],
   "source": [
    "# List of model IDs.\n",
    "models = [\"MGI:3587188\",\"MGI:3587185\",\"MGI:3605874\",\"MGI:2668213\"]\n",
    "\n",
    "# Call iterator function\n",
    "iterator_solr_request(\n",
    "    core='phenodigm', \n",
    "        params = {\n",
    "        'q': 'type:disease_model_summary',  \n",
    "        'fl': 'model_id,marker_id,disease_id',\n",
    "        'field_list': models,\n",
    "        'field_type': 'model_id'\n",
    "    },\n",
    "    filename='model_ids',\n",
    "    format='csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad992401-5ea0-4d16-8777-1a69ce5840b7",
   "metadata": {},
   "source": [
    "### Bonus exercise\n",
    "Now, build your own request based on the example above, incorporating the following changes:\n",
    "1. Use the `genotype-phenotype` core\n",
    "2. Iterate over genes (the name of the field is `marker_symbol`)\n",
    "3. Use the following list of genes: _Zfp580, Firrm, Gpld1, Mbip_\n",
    "4. Modify the list of fields you request (`fl`) to include:\n",
    "    - `marker_symbol`\n",
    "    - `allele_symbol`\n",
    "    - `parameter_stable_id`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3586796f-9cbe-4030-ac26-4282025ebb50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Queried field: marker_symbol:(\"Zfp580\" OR \"Firrm\" OR \"Gpld1\" OR \"Mbip\")\n",
      "Your first request: https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=%2A%3A%2A&fl=marker_symbol%2Callele_symbol%2Cparameter_stable_id&fq=marker_symbol%3A%28%22Zfp580%22+OR+%22Firrm%22+OR+%22Gpld1%22+OR+%22Mbip%22%29&wt=json&start=0&rows=5000\n",
      "Your last request: https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select?q=%2A%3A%2A&fl=marker_symbol%2Callele_symbol%2Cparameter_stable_id&fq=marker_symbol%3A%28%22Zfp580%22+OR+%22Firrm%22+OR+%22Gpld1%22+OR+%22Mbip%22%29&wt=json&start=0&rows=5000\n",
      "Number of found documents: 46\n",
      "\n",
      "File marker_symbol.csv was created.\n"
     ]
    }
   ],
   "source": [
    "# Genes example\n",
    "genes = [\"Zfp580\",\"Firrm\",\"Gpld1\",\"Mbip\"]\n",
    "\n",
    "# Initial query parameters\n",
    "params = {\n",
    "    'q': \"*:*\",\n",
    "    'fl': 'marker_symbol,allele_symbol,parameter_stable_id',\n",
    "    'field_list': genes,\n",
    "    'field_type': \"marker_symbol\"\n",
    "}\n",
    "iterator_solr_request(core='genotype-phenotype', params=params, filename='marker_symbol', format ='csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
